<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Pooja Paresh Savla">

<title>03_modulescore</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="03_modulescore_files/libs/clipboard/clipboard.min.js"></script>
<script src="03_modulescore_files/libs/quarto-html/quarto.js"></script>
<script src="03_modulescore_files/libs/quarto-html/popper.min.js"></script>
<script src="03_modulescore_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="03_modulescore_files/libs/quarto-html/anchor.min.js"></script>
<link href="03_modulescore_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="03_modulescore_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="03_modulescore_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="03_modulescore_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="03_modulescore_files/libs/bootstrap/bootstrap-c0367b04c37547644fece4185067e4a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">03_modulescore</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Pooja Paresh Savla </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="overall-purpose" class="level2">
<h2 class="anchored" data-anchor-id="overall-purpose">Overall purpose</h2>
<p>This pipeline evaluates&nbsp;<strong>pathway activation signatures</strong>&nbsp;(STAT, MAPK, and PI3K/mTOR) from single-cell RNA-seq data after oIL-9R stimulation.<br>
It summarizes&nbsp;<strong>module scores</strong>&nbsp;— averaged expression of key genes representing each signaling pathway — and visualizes them as heatmaps to reveal:</p>
<ul>
<li><p>which pathways are&nbsp;<strong>preferentially activated</strong>&nbsp;across cytokine conditions or cell populations,</p></li>
<li><p>how pathway activation&nbsp;<strong>varies between samples</strong>, and</p></li>
<li><p>whether certain&nbsp;<strong>cell types</strong>&nbsp;are especially responsive to IL-9R signaling.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(dplyr))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(purrr))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tibble))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(SeuratObject))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(Seurat))</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(patchwork))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(ggplot2))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(pheatmap))</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tidyr))</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to save plot as PNG</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>save_plot_as_png <span class="ot">&lt;-</span> <span class="cf">function</span>(plot, filename) {</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">png</span>(filename, <span class="at">res =</span> <span class="dv">250</span>, <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">2000</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/demo_data_annotated_biomart.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define gene sets</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>stat_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Stat1"</span>, <span class="st">"Stat3"</span>, <span class="st">"Stat4"</span>, <span class="st">"Stat5a"</span>, <span class="st">"Stat5b"</span>, <span class="st">"Stat6"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>mapk_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Mapk1"</span>, <span class="st">"Mapk3"</span>, <span class="st">"Map2k1"</span>, <span class="st">"Map2k2"</span>, <span class="st">"Fos"</span>, <span class="st">"Jun"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>pi3k_mtor_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Pik3ca"</span>, <span class="st">"Akt1"</span>, <span class="st">"Akt2"</span>, <span class="st">"Mtor"</span>, <span class="st">"Rps6"</span>, <span class="st">"Rheb"</span>, <span class="st">"Tsc1"</span>, <span class="st">"Tsc2"</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Add module scores for each pathway</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>demo_data <span class="ot">&lt;-</span> <span class="fu">AddModuleScore</span>(demo_data, <span class="at">features =</span> <span class="fu">list</span>(stat_genes), <span class="at">name =</span> <span class="st">"STAT_Module"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
ℹ Please use the `layer` argument instead.
ℹ The deprecated feature was likely used in the Seurat package.
  Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>demo_data <span class="ot">&lt;-</span> <span class="fu">AddModuleScore</span>(demo_data, <span class="at">features =</span> <span class="fu">list</span>(mapk_genes), <span class="at">name =</span> <span class="st">"MAPK_Module"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>demo_data <span class="ot">&lt;-</span> <span class="fu">AddModuleScore</span>(demo_data, <span class="at">features =</span> <span class="fu">list</span>(pi3k_mtor_genes), <span class="at">name =</span> <span class="st">"PI3K_Module"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize module scores</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>module_scores <span class="ot">&lt;-</span> demo_data<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">STAT =</span> <span class="fu">mean</span>(STAT_Module1, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">MAPK =</span> <span class="fu">mean</span>(MAPK_Module1, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">PI3K =</span> <span class="fu">mean</span>(PI3K_Module1, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"samples"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="file-1-modulescore_pathway.png" class="level2">
<h2 class="anchored" data-anchor-id="file-1-modulescore_pathway.png">File 1:&nbsp;<code>modulescore_pathway.png</code></h2>
<p><strong>“Pathway Module Scores Across Samples”</strong></p>
<section id="what-it-contains" class="level3">
<h3 class="anchored" data-anchor-id="what-it-contains">What it contains</h3>
<p>This heatmap displays&nbsp;<strong>average pathway scores (STAT, MAPK, PI3K/mTOR)</strong>&nbsp;across&nbsp;<strong>samples</strong>&nbsp;(columns).<br>
Each pathway is represented as one row.<br>
The color gradient (blue → white → red) corresponds to low → intermediate → high mean pathway activity.</p>
</section>
<section id="how-its-generated" class="level3">
<h3 class="anchored" data-anchor-id="how-its-generated">How it’s generated</h3>
<p>Then&nbsp;<code>pheatmap(t(module_scores))</code>&nbsp;transposes the data so&nbsp;<strong>pathways are rows</strong>&nbsp;and&nbsp;<strong>samples are columns.</strong></p>
</section>
<section id="what-its-asking-biologically" class="level3">
<h3 class="anchored" data-anchor-id="what-its-asking-biologically">What it’s asking biologically</h3>
<p>This heatmap helps you see:</p>
<ul>
<li><p>Which cytokine conditions (samples)&nbsp;<strong>strongly activate</strong>&nbsp;each signaling axis.</p></li>
<li><p>Whether IL-9R stimulation predominantly engages the&nbsp;<strong>STAT</strong>&nbsp;pathway (as expected for many interleukin receptors), or if it co-activates&nbsp;<strong>MAPK</strong>&nbsp;or&nbsp;<strong>PI3K/mTOR</strong>&nbsp;signaling.</p></li>
<li><p>How different experimental conditions (e.g., wild type vs knockout, treated vs control) differ in their&nbsp;<strong>downstream transcriptional activation</strong>.</p></li>
</ul>
<p><strong>Example interpretation:</strong><br>
If the “STAT” row is deep red for oIL-9R samples but white or blue for controls, this supports that&nbsp;<strong>IL-9R signaling drives a JAK/STAT transcriptional program</strong>.<br>
If MAPK or PI3K are also activated, it indicates&nbsp;<strong>cross-talk</strong>&nbsp;or&nbsp;<strong>multi-pathway engagement</strong>, relevant for immune-cell differentiation or survival.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>heatmap_1 <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>(module_scores),  <span class="co"># Transpose: pathways as rows, samples as columns</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))(<span class="dv">100</span>),</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Pathway Module Scores Across Samples"</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontsize_row =</span> <span class="dv">12</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontsize_col =</span> <span class="dv">10</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save_plot_as_png</span>(heatmap_1,<span class="st">"modulescore_pathway.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
</section>
</section>
<section id="file-2-heatmap_stat_module1.png" class="level2">
<h2 class="anchored" data-anchor-id="file-2-heatmap_stat_module1.png">File 2:&nbsp;<code>heatmap_STAT_Module1.png</code></h2>
<p><strong>“STAT module by Sample and Cell Type”</strong></p>
<section id="what-it-contains-1" class="level3">
<h3 class="anchored" data-anchor-id="what-it-contains-1">What it contains</h3>
<p>A heatmap of&nbsp;<strong>STAT pathway module scores</strong>&nbsp;(y-axis = cell types, x-axis = samples).<br>
Each cell type’s mean STAT activity is plotted across all samples.</p>
</section>
<section id="biological-reasoning" class="level3">
<h3 class="anchored" data-anchor-id="biological-reasoning">Biological reasoning</h3>
<p>The&nbsp;<strong>STAT family</strong>&nbsp;(Stat1, Stat3, Stat4, Stat5a/b, Stat6) are core transcription factors downstream of cytokine receptors, including IL-9R.<br>
This heatmap shows:</p>
<ul>
<li><p>Which&nbsp;<strong>cell populations</strong>&nbsp;(e.g., T cells, NK cells, epithelial cells) have strong STAT activation in response to IL-9R signaling.</p></li>
<li><p>Whether&nbsp;<strong>only select subsets</strong>&nbsp;(e.g., IL-9-responsive T cells) show upregulation, reflecting&nbsp;<strong>cell-type-specific responsiveness</strong>.</p></li>
</ul>
<p><strong>Biological question addressed:</strong></p>
<blockquote class="blockquote">
<p>Does oIL-9R activate STAT signaling uniformly across the immune landscape, or are certain populations (e.g., Th9-like or memory T cells) more responsive?</p>
</blockquote>
</section>
</section>
<section id="file-3-heatmap_mapk_module1.png" class="level2">
<h2 class="anchored" data-anchor-id="file-3-heatmap_mapk_module1.png">File 3:&nbsp;<code>heatmap_MAPK_Module1.png</code></h2>
<p><strong>“MAPK module by Sample and Cell Type”</strong></p>
<section id="what-it-contains-2" class="level3">
<h3 class="anchored" data-anchor-id="what-it-contains-2">What it contains</h3>
<p>A heatmap of&nbsp;<strong>MAPK pathway activity</strong>&nbsp;(genes: Mapk1/3, Map2k1/2, Fos, Jun) across cell types and samples.<br>
The color scale (white → yellow → red) indicates increasing activation intensity.</p>
</section>
<section id="biological-reasoning-1" class="level3">
<h3 class="anchored" data-anchor-id="biological-reasoning-1">Biological reasoning</h3>
<p>The&nbsp;<strong>MAPK/ERK</strong>&nbsp;pathway mediates cytokine-driven proliferation and stress responses.<br>
In the IL-9R context, this module tells you:</p>
<ul>
<li><p>Whether&nbsp;<strong>oIL-9R</strong>&nbsp;signaling also triggers&nbsp;<strong>MAPK/ERK</strong>&nbsp;activation in addition to STAT, implying&nbsp;<strong>parallel signaling branches</strong>.</p></li>
<li><p>Which cell types rely on&nbsp;<strong>ERK-dependent programs</strong>&nbsp;(e.g., proliferation, differentiation).</p></li>
</ul>
<p><strong>Possible biological insights:</strong></p>
<ul>
<li><p>High MAPK activity in proliferative clusters (e.g., dividing T cells) suggests&nbsp;<strong>IL-9R-induced proliferation</strong>.</p></li>
<li><p>Weak MAPK activity but strong STAT activity indicates&nbsp;<strong>transcriptional rather than proliferative signaling dominance.</strong></p></li>
</ul>
<hr>
</section>
</section>
<section id="file-4-heatmap_pi3k_module1.png" class="level2">
<h2 class="anchored" data-anchor-id="file-4-heatmap_pi3k_module1.png">File 4:&nbsp;<code>heatmap_PI3K_Module1.png</code></h2>
<p><strong>“PI3K/mTOR module by Sample and Cell Type”</strong></p>
<section id="what-it-contains-3" class="level3">
<h3 class="anchored" data-anchor-id="what-it-contains-3">What it contains</h3>
<p>A heatmap of&nbsp;<strong>PI3K/mTOR pathway scores</strong>&nbsp;(genes: Pik3ca, Akt1/2, Mtor, Rps6, Rheb, Tsc1/2) across samples and annotated cell types.</p>
</section>
<section id="biological-reasoning-2" class="level3">
<h3 class="anchored" data-anchor-id="biological-reasoning-2">Biological reasoning</h3>
<p>The&nbsp;<strong>PI3K-Akt-mTOR</strong>&nbsp;axis controls&nbsp;<strong>metabolic reprogramming</strong>,&nbsp;<strong>cell survival</strong>, and&nbsp;<strong>protein synthesis</strong>.<br>
For IL-9R, this pathway may reflect:</p>
<ul>
<li><p><strong>Survival signaling</strong>&nbsp;downstream of receptor engagement.</p></li>
<li><p><strong>Metabolic adaptation</strong>&nbsp;(e.g., glycolytic shift or biosynthetic activation) in IL-9-responsive immune subsets.</p></li>
</ul>
<p><strong>Biological question addressed:</strong></p>
<blockquote class="blockquote">
<p>Does IL-9R signaling promote metabolic activation via the PI3K/mTOR axis, and in which cell populations is this strongest?</p>
</blockquote>
<p><strong>Interpretation:</strong><br>
High PI3K/mTOR scores in effector or activated cell types suggest&nbsp;<strong>enhanced anabolic metabolism</strong>&nbsp;and&nbsp;<strong>cell growth responses</strong>&nbsp;to IL-9R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the module scores to visualize</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>module_scores <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"STAT_Module1"</span>, <span class="st">"MAPK_Module1"</span>, <span class="st">"PI3K_Module1"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to create a heatmap for each module</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>generate_module_heatmap <span class="ot">&lt;-</span> <span class="cf">function</span>(seurat_object, module_col) {</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 1. Summarize score by sample and cell annotation</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  summary_df <span class="ot">&lt;-</span> seurat_object<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples, cell_annotation) <span class="sc">%&gt;%</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarise</span>(<span class="at">score =</span> <span class="fu">mean</span>(.data[[module_col]], <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.groups =</span> <span class="st">'drop'</span>)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 2. Pivot to wide matrix</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  score_matrix <span class="ot">&lt;-</span> summary_df <span class="sc">%&gt;%</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> samples, <span class="at">values_from =</span> score) <span class="sc">%&gt;%</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"cell_annotation"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.matrix</span>()</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 3. Plot heatmap</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>  heatmap <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    score_matrix,</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span>))(<span class="dv">100</span>),</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">main =</span> <span class="fu">paste</span>(module_col, <span class="st">"by Sample and Cell Type"</span>),</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">fontsize_row =</span> <span class="dv">10</span>,</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">fontsize_col =</span> <span class="dv">10</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">save_plot_as_png</span>(heatmap,<span class="fu">paste0</span>(<span class="st">"heatmap_"</span>,module_col,<span class="st">".png"</span>))</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each module score and plot</span></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (module <span class="cf">in</span> module_scores) {</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">generate_module_heatmap</span>(demo_data, module)</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-6-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="conceptual-summary" class="level2">
<h2 class="anchored" data-anchor-id="conceptual-summary">Conceptual summary</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
</colgroup>
<thead>
<tr class="header">
<th>Module</th>
<th>Key genes</th>
<th>Biological role</th>
<th>What the heatmap reveals</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>STAT</strong></td>
<td>Stat1–6</td>
<td>Canonical transcriptional effectors of cytokine signaling</td>
<td>Transcriptional responsiveness to IL-9R</td>
</tr>
<tr class="even">
<td><strong>MAPK</strong></td>
<td>Mapk1/3, Fos, Jun</td>
<td>Mitogenic and stress pathway, proliferation</td>
<td>Whether IL-9R drives ERK activation</td>
</tr>
<tr class="odd">
<td><strong>PI3K/mTOR</strong></td>
<td>Pik3ca, Akt1/2, Mtor, Rps6, Tsc1/2</td>
<td>Growth/metabolism pathway</td>
<td>Metabolic activation and survival signals</td>
</tr>
</tbody>
</table>
</section>
<section id="why-this-analysis-matters" class="level2">
<h2 class="anchored" data-anchor-id="why-this-analysis-matters">Why this analysis matters</h2>
<ul>
<li><p><strong>Mechanistic validation:</strong>&nbsp;Confirms which intracellular cascades IL-9R primarily uses (STAT-biased vs multi-pathway).</p></li>
<li><p><strong>Cell-type specificity:</strong>&nbsp;Identifies which immune subsets are functionally responsive to IL-9.</p></li>
<li><p><strong>Cross-talk mapping:</strong>&nbsp;Reveals overlap between pathways, hinting at coordinated transcriptional and metabolic rewiring.</p></li>
<li><p><strong>Therapeutic insight:</strong>&nbsp;If certain cell types rely on specific pathways (e.g., STAT5 or mTOR), inhibitors could modulate IL-9-driven immune responses or inflammation.</p></li>
</ul>
</section>
<section id="file-5-scaled_heatmap_pstat.png" class="level2">
<h2 class="anchored" data-anchor-id="file-5-scaled_heatmap_pstat.png">File 5:&nbsp;<strong><code>scaled_heatmap_pSTAT.png</code></strong></h2>
<p><strong>Title in plot:</strong>&nbsp;“Scaled (0–100) — pSTAT Signature across Cytokine Receptors”</p>
<p><strong>What’s inside</strong><br>
Rows: pSTAT1, pSTAT3, pSTAT4, pSTAT5a, pSTAT5b, pSTAT6<br>
Columns: your&nbsp;<em>samples</em>&nbsp;(cytokines/conditions)<br>
Each row is min–max scaled to [0, 100] across columns, so you see where each STAT gene is&nbsp;<em>relatively</em>&nbsp;highest within its own row.</p>
<p><strong>What it’s asking biologically</strong><br>
Which conditions maximally engage each STAT axis?<br>
Typical cues: IFNs → STAT1; IL-6 family/IL-21 → STAT3; IL-12 → STAT4; γc cytokines (IL-2/7/9/15/21) → STAT5; IL-4/13 → STAT6. Peaks suggest ligand–receptor biasing along these axes.</p>
<p><strong>When to trust vs.&nbsp;be cautious</strong></p>
<ul>
<li><p><strong>Useful for:</strong>&nbsp;pattern recognition per gene; ranking which&nbsp;<em>condition</em>&nbsp;best activates a given STAT transcript program.</p></li>
<li><p><strong>Be cautious:</strong>&nbsp;not for comparing&nbsp;<em>magnitudes between genes</em>&nbsp;(scaling removes gene-to-gene baseline differences).</p></li>
</ul>
<p><strong>Why keep both views?</strong><br>
Scaled view highlights&nbsp;<em>which condition</em>&nbsp;is the top hit per gene, independent of absolute expression.</p>
</section>
<section id="file-6-unscaled_avgexp_heatmap_pstat.png" class="level2">
<h2 class="anchored" data-anchor-id="file-6-unscaled_avgexp_heatmap_pstat.png">File 6:&nbsp;<code>unscaled_avgexp_heatmap_pSTAT.png</code></h2>
<p><strong>Title in plot:</strong>&nbsp;“pSTAT Signature across Cytokine Receptors”</p>
<section id="whats-inside" class="level3">
<h3 class="anchored" data-anchor-id="whats-inside">What’s inside</h3>
<ul>
<li>Same rows/columns as above, but values are the&nbsp;<strong>raw average expression</strong>&nbsp;from&nbsp;<code>AverageExpression()</code>&nbsp;(not scaled).</li>
</ul>
</section>
<section id="what-its-asking-biologically-1" class="level3">
<h3 class="anchored" data-anchor-id="what-its-asking-biologically-1">What it’s asking biologically</h3>
<ul>
<li><strong>How strongly expressed are STAT transcripts per condition, in absolute terms?</strong><br>
If&nbsp;<code>STAT5a/b</code>&nbsp;are both high in IL-9R samples relative to control, that’s consistent with a&nbsp;<strong>STAT5-driven</strong>&nbsp;program at the transcriptional level.</li>
</ul>
<p><strong>When to trust vs.&nbsp;be cautious</strong></p>
<ul>
<li><p><strong>Useful for:</strong>&nbsp;magnitude comparisons&nbsp;<em>between</em>&nbsp;genes and conditions; showing real effect size.</p></li>
<li><p><strong>Be cautious:</strong>&nbsp;gene-length/GC/UMI depth and cell-type composition can affect raw averages; interpret alongside QC and cell-state proportions.</p></li>
</ul>
<p><strong>Why keep both views?</strong><br>
Unscaled preserves absolute differences; paired with the scaled view, you get both magnitude and pattern.<strong>Unscaled</strong>&nbsp;= preserves absolute differences (gene A really is higher than gene B).</p>
<ul>
<li><strong>Scaled</strong>&nbsp;= highlights&nbsp;<strong>which condition</strong>&nbsp;is “best” for each gene, irrespective of gene-to-gene baseline.</li>
</ul>
<hr>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Settings ---</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(demo_data) <span class="ot">&lt;-</span> <span class="st">"samples"</span>  <span class="co"># set your grouping</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>pal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"purple"</span>))(<span class="dv">100</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Helpers ---</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>avg_mat_for <span class="ot">&lt;-</span> <span class="cf">function</span>(obj, genes, rename_map, <span class="at">assay =</span> <span class="st">"RNA"</span>) {</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  genes_present <span class="ot">&lt;-</span> <span class="fu">intersect</span>(genes, <span class="fu">rownames</span>(obj[[assay]]))</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(genes_present) <span class="sc">==</span> <span class="dv">0</span>) <span class="fu">stop</span>(<span class="st">"None of the requested genes are in the assay."</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">&lt;-</span> <span class="fu">AverageExpression</span>(</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    obj, <span class="at">assays =</span> assay, <span class="at">features =</span> genes_present, <span class="at">return.seurat =</span> <span class="cn">FALSE</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  )[[assay]]</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># rename rows if a mapping is provided</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  rn <span class="ot">&lt;-</span> <span class="fu">rownames</span>(mat)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  rn2 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(rn <span class="sc">%in%</span> <span class="fu">names</span>(rename_map), rename_map[rn], rn)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> rn2</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>(mat)</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>scale_rows_0_100 <span class="ot">&lt;-</span> <span class="cf">function</span>(mat) {</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>  s <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(mat, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>    rng <span class="ot">&lt;-</span> <span class="fu">range</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">diff</span>(rng) <span class="sc">==</span> <span class="dv">0</span>) <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">length</span>(x)) <span class="cf">else</span> <span class="dv">100</span> <span class="sc">*</span> (x <span class="sc">-</span> rng[<span class="dv">1</span>]) <span class="sc">/</span> <span class="fu">diff</span>(rng)</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>  }))</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># keep dimnames</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(s) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(mat); <span class="fu">colnames</span>(s) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(mat)</span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>  s</span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>pheatmap_create <span class="ot">&lt;-</span> <span class="cf">function</span>(mat, title, <span class="at">palette =</span> pal) {</span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>  p1 <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>    mat,</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> palette,</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">fontsize_row =</span> <span class="dv">12</span>, <span class="at">fontsize_col =</span> <span class="dv">12</span>,</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">angle_col =</span> <span class="dv">45</span>,</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">display_numbers =</span> <span class="fu">round</span>(mat, <span class="dv">2</span>),</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">number_color =</span> <span class="st">"black"</span>,</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">border_color =</span> <span class="st">"grey80"</span>,</span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">main =</span> title,</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">treeheight_row =</span> <span class="dv">0</span>, <span class="at">treeheight_col =</span> <span class="dv">0</span></span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a><span class="co"># =============== pSTAT set ===============</span></span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a>pstat_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Stat1"</span>,<span class="st">"Stat3"</span>,<span class="st">"Stat4"</span>,<span class="st">"Stat5a"</span>,<span class="st">"Stat5b"</span>,<span class="st">"Stat6"</span>)</span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a>pstat_map <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>  <span class="at">Stat1 =</span> <span class="st">"pSTAT1"</span>, <span class="at">Stat3 =</span> <span class="st">"pSTAT3"</span>, <span class="at">Stat4 =</span> <span class="st">"pSTAT4"</span>,</span>
<span id="cb11-49"><a href="#cb11-49" aria-hidden="true" tabindex="-1"></a>  <span class="at">Stat5a =</span> <span class="st">"pSTAT5a"</span>, <span class="at">Stat5b =</span> <span class="st">"pSTAT5b"</span>, <span class="at">Stat6 =</span> <span class="st">"pSTAT6"</span></span>
<span id="cb11-50"><a href="#cb11-50" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-51"><a href="#cb11-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-52"><a href="#cb11-52" aria-hidden="true" tabindex="-1"></a>pstat_avg <span class="ot">&lt;-</span> <span class="fu">avg_mat_for</span>(demo_data, pstat_genes, pstat_map, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.
First group.by variable `ident` starts with a number, appending `g` to ensure valid variable names
This message is displayed once per session.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pstat_scaled <span class="ot">&lt;-</span> <span class="fu">scale_rows_0_100</span>(pstat_avg)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Unscaled</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">pheatmap_create</span>(pstat_avg,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"pSTAT Signature across Cytokine Receptors"</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save_plot_as_png</span>(p1, <span class="st">"unscaled_avgexp_heatmap_pSTAT.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Scaled 0–100</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">pheatmap_create</span>(pstat_scaled,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Scaled (0–100) — pSTAT Signature across Cytokine Receptors"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">save_plot_as_png</span>(p2, <span class="st">"scaled_heatmap_pSTAT.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="file-7-scaled_heatmap_nonstat.png" class="level2">
<h2 class="anchored" data-anchor-id="file-7-scaled_heatmap_nonstat.png">File 7:&nbsp;<code>scaled_heatmap_nonSTAT.png</code></h2>
<p><strong>Title in plot:</strong>&nbsp;“Scaled (0–100) — Non-STAT Signature across Cytokine Receptors”</p>
<section id="whats-inside-1" class="level3">
<h3 class="anchored" data-anchor-id="whats-inside-1">What’s inside</h3>
<ul>
<li>Rows: pErk1 (Mapk3), pErk2 (Mapk1), pAkt1 (Akt1), pAkt2 (Akt2), pAkt3 (Akt3), pS6 (Rps6)<br>
Columns: your&nbsp;<em>samples</em>&nbsp;(cytokines/conditions)<br>
Row-wise min–max scaling to [0, 100] emphasizes&nbsp;<em>relative peaks</em>&nbsp;per pathway component.</li>
</ul>
</section>
<section id="what-its-asking-biologically-2" class="level3">
<h3 class="anchored" data-anchor-id="what-its-asking-biologically-2">What it’s asking biologically</h3>
<ul>
<li>Which conditions preferentially engage the MAPK (ERK1/2), PI3K–Akt (Akt1/2/3), and mTOR (S6) axes? Cytokine receptors can co-activate these pathways via JAK-dependent adapters or parallel signaling; peaks can reveal biased engagement (e.g., strong pS6 suggests mTOR output; strong pErk suggests MAPK bias).</li>
</ul>
<p><strong>When to trust vs.&nbsp;be cautious</strong></p>
<ul>
<li><p><strong>Useful for:</strong>&nbsp;identifying&nbsp;<em>which condition</em>&nbsp;most up-regulates each pathway readout relative to that gene’s own baseline.</p></li>
<li><p><strong>Be cautious:</strong>&nbsp;scaling obscures whether pS6 is globally higher than pErk, etc.—use the unscaled plot to judge magnitude.</p></li>
</ul>
<p><strong>Why keep both views?</strong><br>
Scaled view is for&nbsp;<em>pattern discovery</em>&nbsp;(who peaks where), complementary to the unscaled magnitude view.</p>
</section>
</section>
<section id="file-8-unscaled_avgexp_heatmap_nonstat.png" class="level2">
<h2 class="anchored" data-anchor-id="file-8-unscaled_avgexp_heatmap_nonstat.png">File 8:&nbsp;<code>unscaled_avgexp_heatmap_nonSTAT.png</code></h2>
<p><strong>Title in plot:</strong>&nbsp;“Non-STAT Signature across Cytokine Receptors”</p>
<section id="whats-inside-2" class="level3">
<h3 class="anchored" data-anchor-id="whats-inside-2">What’s inside</h3>
<ul>
<li>Same rows/columns, but raw average expression (no scaling).</li>
</ul>
</section>
<section id="what-its-asking-biologically-3" class="level3">
<h3 class="anchored" data-anchor-id="what-its-asking-biologically-3">What it’s asking biologically</h3>
<ul>
<li>What is the&nbsp;<em>absolute</em>&nbsp;expression level of ERK/Akt/S6 transcripts per condition? This helps judge whether a pathway is broadly higher across conditions (e.g., uniformly elevated Rps6 → ribosomal/mTOR program) versus condition-specific spikes.</li>
</ul>
<p><strong>When to trust vs.&nbsp;be cautious</strong></p>
<ul>
<li><p><strong>Useful for:</strong>&nbsp;real effect sizes across both genes and conditions; comparing MAPK vs PI3K–Akt vs mTOR axes directly.</p></li>
<li><p><strong>Be cautious:</strong>&nbsp;transcript abundance is an&nbsp;<em>indirect</em>&nbsp;proxy of pathway activity—post-translational phosphorylation drives signaling. Interpret with phosphorylation/protein data if available.</p></li>
</ul>
<p><strong>Why keep both views?</strong><br>
Unscaled shows magnitude; together with the scaled plot you get both “how much” and “where it peaks.”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># =============== non-STAT set ===============</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>non_pstat_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Mapk3"</span>,<span class="st">"Mapk1"</span>,<span class="st">"Akt1"</span>,<span class="st">"Akt2"</span>,<span class="st">"Akt3"</span>,<span class="st">"Rps6"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>non_pstat_map <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Mapk3 =</span> <span class="st">"pErk1"</span>, <span class="at">Mapk1 =</span> <span class="st">"pErk2"</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Akt1 =</span> <span class="st">"pAkt1"</span>, <span class="at">Akt2 =</span> <span class="st">"pAkt2"</span>, <span class="at">Akt3 =</span> <span class="st">"pAkt3"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rps6 =</span> <span class="st">"pS6"</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>nonp_avg <span class="ot">&lt;-</span> <span class="fu">avg_mat_for</span>(demo_data, non_pstat_genes, non_pstat_map, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>nonp_scaled <span class="ot">&lt;-</span> <span class="fu">scale_rows_0_100</span>(nonp_avg)</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Unscaled</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">pheatmap_create</span>(nonp_avg,</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a> <span class="st">"Non-STAT Signature across Cytokine Receptors"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save_plot_as_png</span>(p3, <span class="st">"unscaled_avgexp_heatmap_nonSTAT.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Scaled 0–100</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">pheatmap_create</span>(nonp_scaled,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Scaled (0–100) — Non-STAT Signature across Cytokine Receptors"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_modulescore_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save_plot_as_png</span>(p4, <span class="st">"scaled_heatmap_nonSTAT.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
</section>
</section>
<section id="biological-context-caveats" class="level2">
<h2 class="anchored" data-anchor-id="biological-context-caveats">Biological context &amp; caveats</h2>
<ul>
<li><p>These are&nbsp;<strong>STAT gene transcripts</strong>, not phospho-STAT proteins. True pathway activation is phosphorylation and nuclear translocation; RNA captures&nbsp;<strong>downstream transcriptional consequences</strong>&nbsp;and/or&nbsp;<strong>feedback expression</strong>&nbsp;of STATs.</p></li>
<li><p>If you want to directly reflect signaling activity, consider:</p>
<ul>
<li><p>A curated&nbsp;<strong>target-gene signature</strong>&nbsp;for STAT pathways (e.g., known STAT5 targets) rather than STAT transcripts themselves.</p></li>
<li><p>If you have CITE-seq or phospho-flow/Ab-seq: include protein/phospho features (e.g.,&nbsp;<code>pSTAT5</code>,&nbsp;<code>pERK</code>,&nbsp;<code>pAKT</code>,&nbsp;<code>pS6</code>) and compute module scores from those assays.</p></li>
</ul></li>
</ul>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>