<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>06_synergy_demo</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="06_synergy_demo_files/libs/clipboard/clipboard.min.js"></script>
<script src="06_synergy_demo_files/libs/quarto-html/quarto.js"></script>
<script src="06_synergy_demo_files/libs/quarto-html/popper.min.js"></script>
<script src="06_synergy_demo_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="06_synergy_demo_files/libs/quarto-html/anchor.min.js"></script>
<link href="06_synergy_demo_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="06_synergy_demo_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="06_synergy_demo_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="06_synergy_demo_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="06_synergy_demo_files/libs/bootstrap/bootstrap-c0367b04c37547644fece4185067e4a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">06_synergy_demo</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(dplyr))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(purrr))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tibble))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(SeuratObject))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(Seurat))</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(patchwork))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(ggplot2))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(readxl))</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(plyr))</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to save plot as PNG</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>save_plot_as_png <span class="ot">&lt;-</span> <span class="cf">function</span>(plot, filename) {</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">png</span>(filename, <span class="at">res =</span> <span class="dv">250</span>, <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">2000</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load("../data/demo_data_annotated_nobiomart.RData") loading this later</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load("../data/demo_data_annotated_biomart.RData")</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/demo_data_annotated_nobiomart.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Differential expression vs control (10R):</strong><br>
For each condition (<code>12R</code>,&nbsp;<code>4R</code>,&nbsp;<code>12R+4R</code>,&nbsp;<code>4R+12R</code>), we compute DE genes against&nbsp;<strong>10R</strong>&nbsp;with Seurat’s&nbsp;<code>FindMarkers</code>. Significant results are exported at FDR 5% into&nbsp;<code>padj_005/</code>&nbsp;as:</p>
<ul>
<li><code>*_upsig.csv</code>&nbsp;(upregulated),&nbsp;<code>*_downsig.csv</code>&nbsp;(downregulated), and&nbsp;<code>*_sig.csv</code>&nbsp;(all significant), split&nbsp;<strong>per condition</strong>.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>all_combinations <span class="ot">&lt;-</span> <span class="fu">unique</span>(demo_data<span class="sc">$</span>samples)[<span class="sc">!</span><span class="fu">unique</span>(demo_data<span class="sc">$</span>samples) <span class="sc">%in%</span>  <span class="fu">c</span>(<span class="st">"10R"</span>)]</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(all_combinations)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] box    12R    12R+4R 4R     4R+12R
Levels: 12R 10R 12R+4R 4R 4R+12R box</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>find_combined_markers_singletreatment <span class="ot">&lt;-</span> <span class="cf">function</span>(seurat_object, all_combinations,<span class="at">group_column =</span> <span class="st">"sample"</span>,<span class="at">control_name =</span> <span class="st">"PBS"</span>) {</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># seurat_nk_cells need to be a seurat object</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">DefaultAssay</span>(seurat_object) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  cluster_results <span class="ot">&lt;-</span> <span class="fu">lapply</span>(all_combinations, <span class="cf">function</span>(cluster_id) {</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_object,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                           <span class="at">ident.1 =</span> cluster_id,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                           <span class="at">ident.2 =</span> control_name,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                           <span class="at">group.by =</span> group_column,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                           <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">Comparison =</span> <span class="fu">paste</span>(cluster_id, <span class="st">"vs."</span> , <span class="fu">paste0</span>(control_name)),</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">updated_condition =</span> cluster_id</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>      ) </span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">as.data.frame</span>(markers))</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  combined_markers <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, cluster_results)</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(combined_markers)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>all_markers <span class="ot">&lt;-</span> <span class="fu">find_combined_markers_singletreatment</span>(demo_data,all_combinations, <span class="at">group_column =</span> <span class="st">"samples"</span>,<span class="at">control_name =</span> <span class="st">"10R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
ℹ Please use the `layer` argument instead.
ℹ The deprecated feature was likely used in the Seurat package.
  Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `PackageCheck()` was deprecated in SeuratObject 5.0.0.
ℹ Please use `rlang::check_installed()` instead.
ℹ The deprecated feature was likely used in the Seurat package.
  Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(all_markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           gene_name p_val avg_log2FC pct.1 pct.2 p_val_adj  Comparison
1 ENSMUSG00000026950     0  -5.570649 0.103 0.831         0 box vs. 10R
2 ENSMUSG00000099413     0  -3.819130 0.197 0.922         0 box vs. 10R
3 ENSMUSG00000028270     0  -5.242255 0.239 0.958         0 box vs. 10R
4 ENSMUSG00000028268     0  -3.342819 0.128 0.781         0 box vs. 10R
5 ENSMUSG00000079362     0  -3.171229 0.209 0.845         0 box vs. 10R
6 ENSMUSG00000071068     0  -2.805234 0.288 0.910         0 box vs. 10R
  updated_condition
1               box
2               box
3               box
4               box
5               box
6               box</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>subset_genes <span class="ot">&lt;-</span> <span class="cf">function</span>(df, <span class="at">p_val_adj_threshold =</span> <span class="fl">0.05</span>, <span class="at">project_folder =</span> <span class="st">"DEG_exports"</span>) {</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Step 1: Subset genes</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  upregulated <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> p_val_adj <span class="sc">&lt;</span> p_val_adj_threshold)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  downregulated <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(avg_log2FC <span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">0.5</span> <span class="sc">&amp;</span> p_val_adj <span class="sc">&lt;</span> p_val_adj_threshold)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  significant <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">abs</span>(avg_log2FC) <span class="sc">&gt;</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> p_val_adj <span class="sc">&lt;</span> p_val_adj_threshold)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Step 2: Prepare folder</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  padj_name <span class="ot">&lt;-</span> <span class="fu">sprintf</span>(<span class="st">"%03d"</span>, <span class="fu">as.numeric</span>(<span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="st">""</span>, <span class="fu">as.character</span>(p_val_adj_threshold <span class="sc">*</span> <span class="dv">100</span>))))</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># output_folder &lt;- file.path(project_folder, paste0("padj_", padj_name))</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  output_folder <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">paste0</span>(<span class="st">"padj_"</span>, padj_name))</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(output_folder, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Step 3: Save combined CSVs</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(upregulated, <span class="at">file =</span> <span class="fu">file.path</span>(output_folder, <span class="st">"upsigDEGs.csv"</span>), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(downregulated, <span class="at">file =</span> <span class="fu">file.path</span>(output_folder, <span class="st">"downsigDEGs.csv"</span>), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(significant, <span class="at">file =</span> <span class="fu">file.path</span>(output_folder, <span class="st">"allsigDEGs.csv"</span>), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Step 4: Helper to split by updated_condition and export</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>  split_sort_save <span class="ot">&lt;-</span> <span class="cf">function</span>(df, category_suffix) {</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>    df_list <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">split</span>(.<span class="sc">$</span>updated_condition) <span class="sc">%&gt;%</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">map</span>(<span class="sc">~</span> <span class="fu">arrange</span>(.x, <span class="fu">desc</span>(avg_log2FC)))</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Save each condition's file</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">walk2</span>(df_list, <span class="fu">names</span>(df_list), <span class="cf">function</span>(data, condition) {</span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>      safe_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">gsub</span>(<span class="st">"[^a-zA-Z0-9]"</span>, <span class="st">"_"</span>, condition), <span class="st">"_"</span>, category_suffix)</span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">write.csv</span>(data, <span class="at">file =</span> <span class="fu">file.path</span>(output_folder, <span class="fu">paste0</span>(safe_name, <span class="st">".csv"</span>)), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df_list)</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Step 5: Split, save, and return all in a structured list</span></span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">combined =</span> <span class="fu">list</span>(</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>      <span class="at">upregulated =</span> upregulated,</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>      <span class="at">downregulated =</span> downregulated,</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a>      <span class="at">significant =</span> significant</span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">per_condition =</span> <span class="fu">list</span>(</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>      <span class="at">upsig =</span> <span class="fu">split_sort_save</span>(upregulated, <span class="st">"upsig"</span>),</span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>      <span class="at">downsig =</span> <span class="fu">split_sort_save</span>(downregulated, <span class="st">"downsig"</span>),</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>      <span class="at">sig =</span> <span class="fu">split_sort_save</span>(significant, <span class="st">"sig"</span>)</span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>significant_markers <span class="ot">&lt;-</span> <span class="fu">subset_genes</span>(all_markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>folder_path <span class="ot">&lt;-</span> <span class="st">"padj_005"</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>csv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> folder_path, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>dataframes <span class="ot">&lt;-</span> <span class="fu">lapply</span>(csv_files, read.csv)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dataframes) <span class="ot">&lt;-</span> <span class="fu">basename</span>(csv_files)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (name <span class="cf">in</span> <span class="fu">head</span>(<span class="fu">names</span>(dataframes))) {</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"----"</span>, name, <span class="st">"----</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(dataframes[[name]]))</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>---- 10R_downsig.csv ----
[1] gene_name         p_val             avg_log2FC        pct.1            
[5] pct.2             p_val_adj         Comparison        updated_condition
&lt;0 rows&gt; (or 0-length row.names)

---- 10R_sig.csv ----
[1] gene_name         p_val             avg_log2FC        pct.1            
[5] pct.2             p_val_adj         Comparison        updated_condition
&lt;0 rows&gt; (or 0-length row.names)

---- 10R_upsig.csv ----
[1] gene_name         p_val             avg_log2FC        pct.1            
[5] pct.2             p_val_adj         Comparison        updated_condition
&lt;0 rows&gt; (or 0-length row.names)

---- 12R_4R_downsig.csv ----
           gene_name        p_val avg_log2FC pct.1 pct.2    p_val_adj
1 ENSMUSG00000044465 2.690802e-07 -0.5065595 0.356 0.389 2.106844e-02
2 ENSMUSG00000028970 1.708320e-09 -0.5069923 0.385 0.421 1.337580e-04
3 ENSMUSG00000034593 2.344895e-16 -0.5074715 0.638 0.645 1.836006e-11
4 ENSMUSG00000047415 6.285434e-13 -0.5094122 0.497 0.522 4.921369e-08
5 ENSMUSG00000041132 3.091358e-10 -0.5098735 0.419 0.456 2.420471e-05
6 ENSMUSG00000021366 2.089920e-08 -0.5107034 0.519 0.539 1.636365e-03
      Comparison updated_condition
1 12R+4R vs. 10R            12R+4R
2 12R+4R vs. 10R            12R+4R
3 12R+4R vs. 10R            12R+4R
4 12R+4R vs. 10R            12R+4R
5 12R+4R vs. 10R            12R+4R
6 12R+4R vs. 10R            12R+4R

---- 12R_4R_sig.csv ----
           gene_name        p_val avg_log2FC pct.1 pct.2    p_val_adj
1 ENSMUSG00000022464 1.854980e-29   5.442332 0.046 0.001 1.452412e-24
2 ENSMUSG00000020000 2.100316e-59   5.043299 0.099 0.004 1.644505e-54
3 ENSMUSG00000031722 9.730359e-12   4.590123 0.017 0.000 7.618677e-07
4 ENSMUSG00000074607 9.615023e-51   4.094549 0.093 0.007 7.528371e-46
5 ENSMUSG00000111290 1.274793e-08   4.072587 0.011 0.000 9.981377e-04
6 ENSMUSG00000030209 1.055447e-24   3.949002 0.040 0.001 8.263936e-20
      Comparison updated_condition
1 12R+4R vs. 10R            12R+4R
2 12R+4R vs. 10R            12R+4R
3 12R+4R vs. 10R            12R+4R
4 12R+4R vs. 10R            12R+4R
5 12R+4R vs. 10R            12R+4R
6 12R+4R vs. 10R            12R+4R

---- 12R_4R_upsig.csv ----
           gene_name        p_val avg_log2FC pct.1 pct.2    p_val_adj
1 ENSMUSG00000022464 1.854980e-29   5.442332 0.046 0.001 1.452412e-24
2 ENSMUSG00000020000 2.100316e-59   5.043299 0.099 0.004 1.644505e-54
3 ENSMUSG00000031722 9.730359e-12   4.590123 0.017 0.000 7.618677e-07
4 ENSMUSG00000074607 9.615023e-51   4.094549 0.093 0.007 7.528371e-46
5 ENSMUSG00000111290 1.274793e-08   4.072587 0.011 0.000 9.981377e-04
6 ENSMUSG00000030209 1.055447e-24   3.949002 0.040 0.001 8.263936e-20
      Comparison updated_condition
1 12R+4R vs. 10R            12R+4R
2 12R+4R vs. 10R            12R+4R
3 12R+4R vs. 10R            12R+4R
4 12R+4R vs. 10R            12R+4R
5 12R+4R vs. 10R            12R+4R
6 12R+4R vs. 10R            12R+4R</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># treatment_conditions_csv &lt;- list(</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   "12R_4R" = c("12R_4R_sig.csv","12R_sig.csv","4R_sig.csv"),</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   "4R_12R" = c("4R_12R_sig.csv","4R_sig.csv","12R_sig.csv")</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 12R    10R    12R+4R 4R     4R+12R, box</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>treatment_conditions_csv <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"12R_4R"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"12R_4R_upsig.csv"</span>,<span class="st">"12R_upsig.csv"</span>,<span class="st">"4R_upsig.csv"</span>),</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"4R_12R"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"4R_12R_upsig.csv"</span>,<span class="st">"4R_upsig.csv"</span>,<span class="st">"12R_upsig.csv"</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all CSV files in that folder</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>csv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> folder_path, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">upsig.csv$"</span>, <span class="at">full.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co">#🔁 Define renaming patterns</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>rename_patterns <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"_upsig</span><span class="sc">\\</span><span class="st">.csv$"</span> <span class="ot">=</span> <span class="st">".csv"</span>,    <span class="co"># Rename *_sig.csv → .csv</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"_filtered</span><span class="sc">\\</span><span class="st">.csv$"</span> <span class="ot">=</span> <span class="st">"_f.csv"</span>  <span class="co"># (example) Rename *_filtered.csv → *_f.csv</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>treatment_conditions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"12R_4R"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"12R+4R"</span>, <span class="st">"12R"</span>, <span class="st">"4R"</span>),</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"4R_12R"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"4R+12R"</span>, <span class="st">"4R"</span>, <span class="st">"12R"</span>)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="co"># 🛠️ Function to apply renaming rules</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>apply_renames <span class="ot">&lt;-</span> <span class="cf">function</span>(filename, patterns) {</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (pattern <span class="cf">in</span> <span class="fu">names</span>(patterns)) {</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    replacement <span class="ot">&lt;-</span> patterns[[pattern]]</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    filename <span class="ot">&lt;-</span> <span class="fu">sub</span>(pattern, replacement, filename)</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(filename)</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a><span class="co"># 📦 Function to move and rename files into synergy/ folders</span></span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>move_files_to_synergy <span class="ot">&lt;-</span> <span class="cf">function</span>(target_folder, files) {</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">grepl</span>(<span class="st">"^PBS"</span>, target_folder)) {</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"[SKIPPING] Folder starts with 'PBS':"</span>, target_folder, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>()</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>  target_folder_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"synergy"</span>, target_folder)</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(target_folder_path)) {</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(target_folder_path, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (file <span class="cf">in</span> files) {</span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>    original_file_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(folder_path, file)</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">file.exists</span>(original_file_path)) {</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>      output_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(target_folder_path, file)</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a>      <span class="fu">file.copy</span>(original_file_path, output_path, <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"[MOVED] File:"</span>, file, <span class="st">"-&gt;"</span>, output_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"[MISSING] File not found:"</span>, file, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-50"><a href="#cb16-50" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-51"><a href="#cb16-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-52"><a href="#cb16-52" aria-hidden="true" tabindex="-1"></a><span class="co"># 🚚 Iterate over treatment folders and copy files</span></span>
<span id="cb16-53"><a href="#cb16-53" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (folder_name <span class="cf">in</span> <span class="fu">names</span>(treatment_conditions_csv)) {</span>
<span id="cb16-54"><a href="#cb16-54" aria-hidden="true" tabindex="-1"></a>  required_files <span class="ot">&lt;-</span> treatment_conditions_csv[[folder_name]]</span>
<span id="cb16-55"><a href="#cb16-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">move_files_to_synergy</span>(folder_name, required_files)</span>
<span id="cb16-56"><a href="#cb16-56" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[MOVED] File: 12R_4R_upsig.csv -&gt; synergy/12R_4R/12R_4R_upsig.csv 
[MOVED] File: 12R_upsig.csv -&gt; synergy/12R_4R/12R_upsig.csv 
[MOVED] File: 4R_upsig.csv -&gt; synergy/12R_4R/4R_upsig.csv 
[MOVED] File: 4R_12R_upsig.csv -&gt; synergy/4R_12R/4R_12R_upsig.csv 
[MOVED] File: 4R_upsig.csv -&gt; synergy/4R_12R/4R_upsig.csv 
[MOVED] File: 12R_upsig.csv -&gt; synergy/4R_12R/12R_upsig.csv </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ✅ Summary</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">[INFO] All matching files moved with optional renaming.</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
[INFO] All matching files moved with optional renaming.</code></pre>
</div>
</div>
<p><strong>Curate combination results:</strong><br>
Files are organized into&nbsp;<code>synergy/&lt;pair&gt;/</code>&nbsp;(e.g.,&nbsp;<code>synergy/12R_4R/</code>). For each pair we tabulate:</p>
<ul>
<li><p><strong>common genes</strong>&nbsp;(present across all relevant files),</p></li>
<li><p><strong>unique genes</strong>&nbsp;(present in some but not all),</p></li>
<li><p><strong>exclusive genes</strong>&nbsp;(specific to one file).</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required libraries</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the main directory containing cytokine condition folders</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>synergy_dir <span class="ot">&lt;-</span> <span class="st">"synergy"</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all subfolders inside synergy/</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>subfolders <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(synergy_dir, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over each folder</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (target_folder <span class="cf">in</span> subfolders) {</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">🔍 Processing folder:"</span>, target_folder, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure the folder contains CSV files</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  csv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> target_folder, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(csv_files) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"[WARNING] No CSV files found in:"</span>, target_folder, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">next</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define results subfolder</span></span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>  results_folder <span class="ot">&lt;-</span> <span class="fu">file.path</span>(target_folder, <span class="st">"analysis_results"</span>)</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create results folder if it doesn't exist</span></span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(results_folder)) {</span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(results_folder, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize an empty list to store gene lists from each file</span></span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>  gene_lists <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>  gene_data <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Read each CSV file and extract the "gene_name" column</span></span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (file <span class="cf">in</span> csv_files) {</span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(file, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Ensure "gene_name" column exists</span></span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="st">"gene_name"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(df)) {</span>
<span id="cb20-41"><a href="#cb20-41" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"[WARNING] Missing 'gene_name' column in:"</span>, file, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-42"><a href="#cb20-42" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb20-43"><a href="#cb20-43" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb20-44"><a href="#cb20-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-45"><a href="#cb20-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Store gene names in a list with the filename as key</span></span>
<span id="cb20-46"><a href="#cb20-46" aria-hidden="true" tabindex="-1"></a>    gene_lists[[<span class="fu">basename</span>(file)]] <span class="ot">&lt;-</span> <span class="fu">unique</span>(df<span class="sc">$</span>gene_name)</span>
<span id="cb20-47"><a href="#cb20-47" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-48"><a href="#cb20-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-49"><a href="#cb20-49" aria-hidden="true" tabindex="-1"></a>  <span class="co"># with log2fc</span></span>
<span id="cb20-50"><a href="#cb20-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># With avg_log2FC if available</span></span>
<span id="cb20-51"><a href="#cb20-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"avg_log2FC"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(df)) {</span>
<span id="cb20-52"><a href="#cb20-52" aria-hidden="true" tabindex="-1"></a>      gene_data[[<span class="fu">basename</span>(file)]] <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb20-53"><a href="#cb20-53" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">Gene =</span> gene_name, avg_log2FC) <span class="sc">%&gt;%</span></span>
<span id="cb20-54"><a href="#cb20-54" aria-hidden="true" tabindex="-1"></a>        <span class="fu">distinct</span>()</span>
<span id="cb20-55"><a href="#cb20-55" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb20-56"><a href="#cb20-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-57"><a href="#cb20-57" aria-hidden="true" tabindex="-1"></a>  <span class="co"># If no valid gene data found, skip folder</span></span>
<span id="cb20-58"><a href="#cb20-58" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(gene_lists) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb20-59"><a href="#cb20-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"[WARNING] No valid gene data found in:"</span>, target_folder, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-60"><a href="#cb20-60" aria-hidden="true" tabindex="-1"></a>    <span class="cf">next</span></span>
<span id="cb20-61"><a href="#cb20-61" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-62"><a href="#cb20-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-63"><a href="#cb20-63" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Find shared genes (common across all files)</span></span>
<span id="cb20-64"><a href="#cb20-64" aria-hidden="true" tabindex="-1"></a>  common_genes <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(intersect, gene_lists)</span>
<span id="cb20-65"><a href="#cb20-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-66"><a href="#cb20-66" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Find unique genes (present in at least one file but not all)</span></span>
<span id="cb20-67"><a href="#cb20-67" aria-hidden="true" tabindex="-1"></a>  all_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">unlist</span>(gene_lists))</span>
<span id="cb20-68"><a href="#cb20-68" aria-hidden="true" tabindex="-1"></a>  unique_genes <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(all_genes, common_genes)</span>
<span id="cb20-69"><a href="#cb20-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-70"><a href="#cb20-70" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Find exclusive genes (genes that appear in only one file)</span></span>
<span id="cb20-71"><a href="#cb20-71" aria-hidden="true" tabindex="-1"></a>  exclusive_genes <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-72"><a href="#cb20-72" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (file <span class="cf">in</span> <span class="fu">names</span>(gene_lists)) {</span>
<span id="cb20-73"><a href="#cb20-73" aria-hidden="true" tabindex="-1"></a>    exclusive_genes[[file]] <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(gene_lists[[file]], <span class="fu">unlist</span>(gene_lists[<span class="fu">names</span>(gene_lists) <span class="sc">!=</span> file]))</span>
<span id="cb20-74"><a href="#cb20-74" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-75"><a href="#cb20-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-76"><a href="#cb20-76" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(<span class="fu">data.frame</span>(<span class="at">Gene=</span>common_genes), <span class="fu">file.path</span>(results_folder, <span class="st">"common_genes.csv"</span>), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-77"><a href="#cb20-77" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(<span class="fu">data.frame</span>(<span class="at">Gene=</span>unique_genes), <span class="fu">file.path</span>(results_folder, <span class="st">"unique_genes.csv"</span>), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-78"><a href="#cb20-78" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (file <span class="cf">in</span> <span class="fu">names</span>(exclusive_genes)) {</span>
<span id="cb20-79"><a href="#cb20-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">write.csv</span>(<span class="fu">data.frame</span>(<span class="at">Gene=</span>exclusive_genes[[file]]), <span class="fu">file.path</span>(results_folder, <span class="fu">paste0</span>(<span class="st">"exclusive_genes_"</span>, file)), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-80"><a href="#cb20-80" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-81"><a href="#cb20-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-82"><a href="#cb20-82" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save avg_log2FC-based versions if available</span></span>
<span id="cb20-83"><a href="#cb20-83" aria-hidden="true" tabindex="-1"></a><span class="co"># Save avg_log2FC-based versions if available</span></span>
<span id="cb20-84"><a href="#cb20-84" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(gene_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb20-85"><a href="#cb20-85" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Unique genes with avg_log2FC</span></span>
<span id="cb20-86"><a href="#cb20-86" aria-hidden="true" tabindex="-1"></a>  unique_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb20-87"><a href="#cb20-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(gene_data, <span class="cf">function</span>(df) {</span>
<span id="cb20-88"><a href="#cb20-88" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(df)) df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Gene <span class="sc">%in%</span> unique_genes)</span>
<span id="cb20-89"><a href="#cb20-89" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb20-90"><a href="#cb20-90" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(avg_log2FC))</span>
<span id="cb20-91"><a href="#cb20-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-92"><a href="#cb20-92" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(unique_df, <span class="fu">file.path</span>(results_folder, <span class="st">"unique_genes_with_log2FC.csv"</span>), <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-93"><a href="#cb20-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-94"><a href="#cb20-94" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Exclusive genes with avg_log2FC</span></span>
<span id="cb20-95"><a href="#cb20-95" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (file <span class="cf">in</span> <span class="fu">names</span>(exclusive_genes)) {</span>
<span id="cb20-96"><a href="#cb20-96" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(gene_data[[file]])) {</span>
<span id="cb20-97"><a href="#cb20-97" aria-hidden="true" tabindex="-1"></a>      exclusive_df <span class="ot">&lt;-</span> gene_data[[file]] <span class="sc">%&gt;%</span></span>
<span id="cb20-98"><a href="#cb20-98" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(Gene <span class="sc">%in%</span> exclusive_genes[[file]]) <span class="sc">%&gt;%</span></span>
<span id="cb20-99"><a href="#cb20-99" aria-hidden="true" tabindex="-1"></a>        <span class="fu">arrange</span>(<span class="fu">desc</span>(avg_log2FC))</span>
<span id="cb20-100"><a href="#cb20-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-101"><a href="#cb20-101" aria-hidden="true" tabindex="-1"></a>      <span class="fu">write.csv</span>(</span>
<span id="cb20-102"><a href="#cb20-102" aria-hidden="true" tabindex="-1"></a>        exclusive_df,</span>
<span id="cb20-103"><a href="#cb20-103" aria-hidden="true" tabindex="-1"></a>        <span class="fu">file.path</span>(results_folder, <span class="fu">paste0</span>(<span class="st">"exclusive_genes_with_log2FC_"</span>, file)),</span>
<span id="cb20-104"><a href="#cb20-104" aria-hidden="true" tabindex="-1"></a>        <span class="at">row.names =</span> <span class="cn">FALSE</span></span>
<span id="cb20-105"><a href="#cb20-105" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb20-106"><a href="#cb20-106" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb20-107"><a href="#cb20-107" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-108"><a href="#cb20-108" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-109"><a href="#cb20-109" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print summary</span></span>
<span id="cb20-110"><a href="#cb20-110" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">✅ Comparison completed for:"</span>, target_folder, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-111"><a href="#cb20-111" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"✅ Common genes across all files:"</span>, <span class="fu">length</span>(common_genes), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-112"><a href="#cb20-112" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"✅ Unique genes across some files:"</span>, <span class="fu">length</span>(unique_genes), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-113"><a href="#cb20-113" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"✅ Exclusive genes (by file) saved separately.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-114"><a href="#cb20-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"📂 Results saved in:"</span>, results_folder, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb20-115"><a href="#cb20-115" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
🔍 Processing folder: synergy/12R_4R 

✅ Comparison completed for: synergy/12R_4R 
✅ Common genes across all files: 12 
✅ Unique genes across some files: 5357 
✅ Exclusive genes (by file) saved separately.
📂 Results saved in: synergy/12R_4R/analysis_results 

🔍 Processing folder: synergy/4R_12R 

✅ Comparison completed for: synergy/4R_12R 
✅ Common genes across all files: 12 
✅ Unique genes across some files: 5420 
✅ Exclusive genes (by file) saved separately.
📂 Results saved in: synergy/4R_12R/analysis_results </code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">🚀 **Batch processing completed for all folders in synergy/**</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
🚀 **Batch processing completed for all folders in synergy/**</code></pre>
</div>
</div>
<p>Call Exclusive Genes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the main synergy directory</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>synergy_dir <span class="ot">&lt;-</span> <span class="st">"synergy/"</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to read common genes from CSV</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>read_common_genes <span class="ot">&lt;-</span> <span class="cf">function</span>(file_path) {</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">file.exists</span>(file_path)) {</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(file_path, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"Gene"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(df)) {</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="fu">unique</span>(df<span class="sc">$</span>Gene))  <span class="co"># Return unique common genes</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"[WARNING] Missing 'Gene' column in:"</span>, file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to read common genes from CSV</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>read_exclusive_genes <span class="ot">&lt;-</span> <span class="cf">function</span>(file_path) {</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">file.exists</span>(file_path)) {</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(file_path, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"Gene"</span> <span class="sc">%in%</span> <span class="fu">colnames</span>(df)) {</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="fu">unique</span>(df<span class="sc">$</span>Gene))  <span class="co"># Return unique common genes</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"[WARNING] Missing 'Gene' column in:"</span>, file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all treatment folders inside the synergy directory</span></span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>synergy_folders <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="at">path =</span> synergy_dir, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize a list to store common genes per treatment</span></span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>common_genes_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over each folder inside synergy/</span></span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (treatment_folder <span class="cf">in</span> synergy_folders) {</span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define the common_genes.csv file path</span></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a>  common_genes_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(treatment_folder, <span class="st">"analysis_results"</span>, <span class="st">"common_genes.csv"</span>)</span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract treatment condition from folder name</span></span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a>  treatment_condition <span class="ot">&lt;-</span> <span class="fu">basename</span>(treatment_folder)</span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Read and store common genes if the file exists</span></span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a>  common_genes <span class="ot">&lt;-</span> <span class="fu">read_common_genes</span>(common_genes_file)</span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(common_genes) <span class="sc">&amp;&amp;</span> <span class="fu">length</span>(common_genes) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a>    common_genes_list[[treatment_condition]] <span class="ot">&lt;-</span> common_genes</span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">🔹 **Treatment Condition:**"</span>, treatment_condition, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"📄 **Number of Common Genes:**"</span>, <span class="fu">length</span>(common_genes), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">head</span>(common_genes, <span class="dv">10</span>))  <span class="co"># Show first 10 genes as preview</span></span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">[INFO] No common genes found for:"</span>, treatment_condition, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
🔹 **Treatment Condition:** 12R_4R 
📄 **Number of Common Genes:** 12 
 [1] "ENSMUSG00000061603" "ENSMUSG00000061175" "ENSMUSG00000031618"
 [4] "ENSMUSG00000009418" "ENSMUSG00000021037" "ENSMUSG00000097471"
 [7] "ENSMUSG00000038594" "ENSMUSG00000033114" "ENSMUSG00000035842"
[10] "ENSMUSG00000009828"
...

🔹 **Treatment Condition:** 4R_12R 
📄 **Number of Common Genes:** 12 
 [1] "ENSMUSG00000061603" "ENSMUSG00000061175" "ENSMUSG00000008496"
 [4] "ENSMUSG00000019982" "ENSMUSG00000009418" "ENSMUSG00000035842"
 [7] "ENSMUSG00000021037" "ENSMUSG00000033114" "ENSMUSG00000038594"
[10] "ENSMUSG00000037679"
...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">✅ Process completed. Common genes extracted for each treatment condition.</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
✅ Process completed. Common genes extracted for each treatment condition.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Exclusive</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize a list to store exclusive genes per treatment</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>exclusive_genes_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over each folder inside synergy/</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (treatment_folder <span class="cf">in</span> synergy_folders) {</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define the exclusive genes CSV file path</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  exclusive_genes_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(treatment_folder, <span class="st">"analysis_results"</span>, </span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">paste0</span>(<span class="st">"exclusive_genes_"</span>, <span class="fu">basename</span>(treatment_folder), <span class="st">"_upsig.csv"</span>))</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract treatment condition from folder name</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  treatment_condition <span class="ot">&lt;-</span> <span class="fu">basename</span>(treatment_folder)</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Read and store exclusive genes if the file exists</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>  exclusive_genes <span class="ot">&lt;-</span> <span class="fu">read_exclusive_genes</span>(exclusive_genes_file)</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(exclusive_genes) <span class="sc">&amp;&amp;</span> <span class="fu">length</span>(exclusive_genes) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>    exclusive_genes_list[[treatment_condition]] <span class="ot">&lt;-</span> exclusive_genes</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">🔸 **Treatment Condition:**"</span>, treatment_condition, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"🧬 **Number of Exclusive Genes:**"</span>, <span class="fu">length</span>(exclusive_genes), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">head</span>(exclusive_genes, <span class="dv">10</span>))  <span class="co"># Show first 10 genes as preview</span></span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">[INFO] No exclusive genes found for:"</span>, treatment_condition, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
🔸 **Treatment Condition:** 12R_4R 
🧬 **Number of Exclusive Genes:** 565 
 [1] "ENSMUSG00000111290" "ENSMUSG00000030209" "ENSMUSG00000079157"
 [4] "ENSMUSG00000120934" "ENSMUSG00000135363" "ENSMUSG00000127105"
 [7] "ENSMUSG00000019122" "ENSMUSG00000047414" "ENSMUSG00000003559"
[10] "ENSMUSG00000108490"
...

🔸 **Treatment Condition:** 4R_12R 
🧬 **Number of Exclusive Genes:** 628 
 [1] "ENSMUSG00000130707" "ENSMUSG00000138022" "ENSMUSG00000087641"
 [4] "ENSMUSG00000137281" "ENSMUSG00000086686" "ENSMUSG00000132261"
 [7] "ENSMUSG00000021804" "ENSMUSG00000050931" "ENSMUSG00000038859"
[10] "ENSMUSG00000003436"
...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">✅ Process completed. Exclusive genes extracted for each treatment condition.</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
✅ Process completed. Exclusive genes extracted for each treatment condition.</code></pre>
</div>
</div>
<p><strong>Synergy test (Z &gt; X + Y):</strong><br>
We formalize synergy for a combination&nbsp;<strong>Z = (12R+4R or 4R+12R)</strong>&nbsp;vs its components&nbsp;<strong>X</strong>&nbsp;and&nbsp;<strong>Y</strong>&nbsp;by computing&nbsp;<strong>per-gene mean expression</strong>&nbsp;and flagging&nbsp;<strong>synergy genes</strong>&nbsp;where:</p>
<p>mean(Z)  &gt;  mean(X)+mean(Y).mean(Z)&gt;mean(X)+mean(Y).</p>
<p>This highlights&nbsp;<strong>supra-additive</strong>&nbsp;transcriptional programs attributable to the combined signal, not seen with single receptors alone.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>synergy_dir <span class="ot">&lt;-</span> <span class="st">"synergy/"</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all treatment folders inside the synergy directory</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>synergy_folders <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="at">path =</span> synergy_dir, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize an empty list to store filenames without extensions</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>all_filenames <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over each treatment folder</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (treatment_folder <span class="cf">in</span> synergy_folders) {</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># List all CSV files in the folder</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>  csv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> treatment_folder, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove the ".csv" extension and store filenames</span></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(csv_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>    filenames_no_ext <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">_upsig.csv$"</span>, <span class="st">""</span>, <span class="fu">basename</span>(csv_files))</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>    all_filenames[[<span class="fu">basename</span>(treatment_folder)]] <span class="ot">&lt;-</span> filenames_no_ext</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MAKE SURE THIS REPRESENTS (FOLDER = SAMPLES in a list)</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>treatment_conditions <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"12R_4R"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"12R+4R"</span>,<span class="st">"12R"</span>,<span class="st">"4R"</span>),</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"4R_12R"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"4R+12R"</span>,<span class="st">"4R"</span>,<span class="st">"12R"</span>)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize lists to store synergy genes</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>exclusive_genes_list_synergy <span class="ot">&lt;-</span> <span class="fu">list</span>()  <span class="co"># Z &gt; X + Y (no PBS)</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all treatment folders inside synergy/</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>treatment_folders <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="at">path =</span> synergy_dir, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="co">#treatment_folder &lt;- "synergy//12R_4R"</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each treatment folder</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (treatment_folder <span class="cf">in</span> treatment_folders) {</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>  treatment_condition <span class="ot">&lt;-</span> <span class="fu">basename</span>(treatment_folder)</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>  exclusive_genes <span class="ot">&lt;-</span> exclusive_genes_list[[treatment_condition]]</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Skip if no exclusive genes available</span></span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(exclusive_genes)) {</span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"⚠️ Skipping"</span>, treatment_condition, <span class="st">"- no exclusive genes found.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">next</span></span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define sample names</span></span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>  valid_conditions <span class="ot">&lt;-</span> <span class="fu">c</span>(treatment_conditions[[treatment_condition]], <span class="st">"10R"</span>)</span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a>  z_sample <span class="ot">&lt;-</span> treatment_conditions[[treatment_condition]][<span class="dv">1</span>]  <span class="co"># actual sample name like "21R+2R"</span></span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset Seurat object</span></span>
<span id="cb33-31"><a href="#cb33-31" aria-hidden="true" tabindex="-1"></a>  subset_seurat <span class="ot">&lt;-</span> <span class="fu">subset</span>(demo_data, <span class="at">subset =</span> samples <span class="sc">%in%</span> valid_conditions)</span>
<span id="cb33-32"><a href="#cb33-32" aria-hidden="true" tabindex="-1"></a>  genes_in_seurat <span class="ot">&lt;-</span> <span class="fu">intersect</span>(exclusive_genes, <span class="fu">rownames</span>(subset_seurat))</span>
<span id="cb33-33"><a href="#cb33-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-34"><a href="#cb33-34" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Reorder factor levels</span></span>
<span id="cb33-35"><a href="#cb33-35" aria-hidden="true" tabindex="-1"></a>  subset_seurat<span class="sc">$</span>samples <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb33-36"><a href="#cb33-36" aria-hidden="true" tabindex="-1"></a>    subset_seurat<span class="sc">$</span>samples,</span>
<span id="cb33-37"><a href="#cb33-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"10R"</span>,</span>
<span id="cb33-38"><a href="#cb33-38" aria-hidden="true" tabindex="-1"></a>               <span class="fu">setdiff</span>(<span class="fu">unique</span>(subset_seurat<span class="sc">$</span>samples), <span class="fu">c</span>(<span class="st">"10R"</span>, z_sample)),</span>
<span id="cb33-39"><a href="#cb33-39" aria-hidden="true" tabindex="-1"></a>               z_sample)</span>
<span id="cb33-40"><a href="#cb33-40" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb33-41"><a href="#cb33-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-42"><a href="#cb33-42" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize gene list</span></span>
<span id="cb33-43"><a href="#cb33-43" aria-hidden="true" tabindex="-1"></a>  synergy_genes <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb33-44"><a href="#cb33-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-45"><a href="#cb33-45" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Detect synergy genes</span></span>
<span id="cb33-46"><a href="#cb33-46" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (gene <span class="cf">in</span> genes_in_seurat) {</span>
<span id="cb33-47"><a href="#cb33-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">try</span>({</span>
<span id="cb33-48"><a href="#cb33-48" aria-hidden="true" tabindex="-1"></a>      data <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(subset_seurat, <span class="at">vars =</span> <span class="fu">c</span>(gene, <span class="st">"samples"</span>))</span>
<span id="cb33-49"><a href="#cb33-49" aria-hidden="true" tabindex="-1"></a>      <span class="fu">colnames</span>(data)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"expression"</span></span>
<span id="cb33-50"><a href="#cb33-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-51"><a href="#cb33-51" aria-hidden="true" tabindex="-1"></a>      group_means <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb33-52"><a href="#cb33-52" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb33-53"><a href="#cb33-53" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">mean_expression =</span> <span class="fu">mean</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.groups =</span> <span class="st">"drop"</span>)</span>
<span id="cb33-54"><a href="#cb33-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-55"><a href="#cb33-55" aria-hidden="true" tabindex="-1"></a>      z_mean <span class="ot">&lt;-</span> group_means <span class="sc">%&gt;%</span></span>
<span id="cb33-56"><a href="#cb33-56" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">filter</span>(samples <span class="sc">==</span> z_sample) <span class="sc">%&gt;%</span></span>
<span id="cb33-57"><a href="#cb33-57" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">pull</span>(mean_expression)</span>
<span id="cb33-58"><a href="#cb33-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-59"><a href="#cb33-59" aria-hidden="true" tabindex="-1"></a>      control_means_no_pbs <span class="ot">&lt;-</span> group_means <span class="sc">%&gt;%</span></span>
<span id="cb33-60"><a href="#cb33-60" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">filter</span>(samples <span class="sc">!=</span> z_sample <span class="sc">&amp;</span> samples <span class="sc">!=</span> <span class="st">"10R"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-61"><a href="#cb33-61" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">pull</span>(mean_expression)</span>
<span id="cb33-62"><a href="#cb33-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-63"><a href="#cb33-63" aria-hidden="true" tabindex="-1"></a>      control_sum_no_pbs <span class="ot">&lt;-</span> <span class="fu">sum</span>(control_means_no_pbs, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-64"><a href="#cb33-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-65"><a href="#cb33-65" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">length</span>(z_mean) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> z_mean <span class="sc">&gt;</span> control_sum_no_pbs) {</span>
<span id="cb33-66"><a href="#cb33-66" aria-hidden="true" tabindex="-1"></a>        <span class="co">#cat("✅ [v2]", gene, "| Z mean:", round(z_mean, 2),</span></span>
<span id="cb33-67"><a href="#cb33-67" aria-hidden="true" tabindex="-1"></a>           <span class="co"># "&gt;", "X+Y sum:", round(control_sum_no_pbs, 2), "\n")</span></span>
<span id="cb33-68"><a href="#cb33-68" aria-hidden="true" tabindex="-1"></a>        synergy_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(synergy_genes, gene)</span>
<span id="cb33-69"><a href="#cb33-69" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb33-70"><a href="#cb33-70" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb33-71"><a href="#cb33-71" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb33-72"><a href="#cb33-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-73"><a href="#cb33-73" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save if any found</span></span>
<span id="cb33-74"><a href="#cb33-74" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(synergy_genes) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb33-75"><a href="#cb33-75" aria-hidden="true" tabindex="-1"></a>    exclusive_genes_list_synergy[[treatment_condition]] <span class="ot">&lt;-</span> <span class="fu">unique</span>(synergy_genes)</span>
<span id="cb33-76"><a href="#cb33-76" aria-hidden="true" tabindex="-1"></a>    <span class="co">#cat("🧬 [v2] Saved", length(synergy_genes), "synergy genes for", treatment_condition, "\n")</span></span>
<span id="cb33-77"><a href="#cb33-77" aria-hidden="true" tabindex="-1"></a>  <span class="co">#} else {</span></span>
<span id="cb33-78"><a href="#cb33-78" aria-hidden="true" tabindex="-1"></a>    <span class="co">#cat("❌ [v2] No synergy genes found for:", treatment_condition, "\n")</span></span>
<span id="cb33-79"><a href="#cb33-79" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb33-80"><a href="#cb33-80" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb33-81"><a href="#cb33-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-82"><a href="#cb33-82" aria-hidden="true" tabindex="-1"></a><span class="co"># Final Summary</span></span>
<span id="cb33-83"><a href="#cb33-83" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">🎯 Completed synergy detection for all treatments.</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
🎯 Completed synergy detection for all treatments.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">🔸 Version 2 (Z &gt; X + Y):</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
🔸 Version 2 (Z &gt; X + Y):</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">sapply</span>(exclusive_genes_list_synergy, length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>12R_4R 4R_12R 
   174    139 </code></pre>
</div>
</div>
<p><strong>Interpretation plots (decomposition barplot):</strong><br>
For the top synergy candidates we plot, per gene, a stacked bar with:</p>
<ul>
<li><p><strong>Difference</strong>&nbsp;=&nbsp;Z−X−YZ−X−Y&nbsp;(excess over additivity),</p></li>
<li><p>the&nbsp;<strong>X</strong>&nbsp;and&nbsp;<strong>Y</strong>&nbsp;components.<br>
Gene IDs are mapped from Ensembl →&nbsp;<strong>MGI symbols</strong>&nbsp;for readability.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>plot_synergy_barplot <span class="ot">&lt;-</span> <span class="cf">function</span>(df_long,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                                 treatment_name,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                                 custom_colors,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">sample_col =</span> <span class="st">"Sample"</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">expression_col =</span> <span class="st">"Expression"</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">cytokine_col =</span> <span class="st">"Cytokine"</span>,</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">plot_title_suffix =</span> <span class="st">"Decomposition by Component"</span>) {</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Convert strings to symbols for tidy evaluation</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  sample_sym <span class="ot">&lt;-</span> <span class="fu">sym</span>(sample_col)</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  expression_sym <span class="ot">&lt;-</span> <span class="fu">sym</span>(expression_col)</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  cytokine_sym <span class="ot">&lt;-</span> <span class="fu">sym</span>(cytokine_col)</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure proper factor levels for Sample and Cytokine</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>  df_long[[sample_col]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_long[[sample_col]], <span class="at">levels =</span> <span class="fu">unique</span>(df_long[[sample_col]]))</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>  df_long[[cytokine_col]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_long[[cytokine_col]], <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Difference"</span>, <span class="fu">names</span>(custom_colors)[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>]))</span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Generate stacked bar plot</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> <span class="sc">!!</span>sample_sym, <span class="at">y =</span> <span class="sc">!!</span>expression_sym, <span class="at">fill =</span> <span class="sc">!!</span>cytokine_sym)) <span class="sc">+</span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"stack"</span>, <span class="at">width =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> custom_colors, <span class="at">name =</span> <span class="st">"Component"</span>) <span class="sc">+</span></span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="fu">paste</span>(treatment_name, plot_title_suffix),</span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">subtitle =</span> <span class="st">"Ordered by Total Expression (Treatment + Controls)"</span>,</span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Expression"</span></span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">13</span>) <span class="sc">+</span></span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold"</span>),</span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">16</span>),</span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold"</span>),</span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">11</span>),</span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">0</span>)</span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Interpretation plots (decomposition barplot):</strong><br>
For the top synergy candidates we plot, per gene, a stacked bar with:</p>
<ul>
<li><p><strong>Difference</strong>&nbsp;=&nbsp;Z−X−YZ−X−Y&nbsp;(excess over additivity),</p></li>
<li><p>the&nbsp;<strong>X</strong>&nbsp;and&nbsp;<strong>Y</strong>&nbsp;components.<br>
Gene IDs are mapped from Ensembl →&nbsp;<strong>MGI symbols</strong>&nbsp;for readability.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (treatment <span class="cf">in</span> <span class="fu">names</span>(treatment_conditions)){</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>valid_conditions <span class="ot">&lt;-</span> treatment_conditions[[treatment]]</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign condition names</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>treatment_name <span class="ot">&lt;-</span> valid_conditions[[<span class="dv">1</span>]]</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>control1_name  <span class="ot">&lt;-</span> valid_conditions[[<span class="dv">2</span>]]</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>control2_name  <span class="ot">&lt;-</span> valid_conditions[[<span class="dv">3</span>]]</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset Seurat object</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>subset_seurat <span class="ot">&lt;-</span> <span class="fu">subset</span>(demo_data, <span class="at">subset =</span> samples <span class="sc">%in%</span> valid_conditions)</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Get exclusive genes that are also in the Seurat object</span></span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>genes_in_seurat <span class="ot">&lt;-</span> <span class="fu">intersect</span>(exclusive_genes_list[[treatment]], <span class="fu">rownames</span>(subset_seurat))</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Order factor levels for consistency</span></span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>subset_seurat<span class="sc">$</span>samples <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a>  subset_seurat<span class="sc">$</span>samples,</span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"10R"</span>, <span class="fu">setdiff</span>(valid_conditions, <span class="st">"10R"</span>))</span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a>expr_data <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(subset_seurat, <span class="at">vars =</span> <span class="fu">c</span>(genes_in_seurat, <span class="st">"samples"</span>))</span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a>summary_df <span class="ot">&lt;-</span> expr_data <span class="sc">%&gt;%</span></span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">all_of</span>(genes_in_seurat), <span class="sc">~</span><span class="fu">mean</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>()</span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-33"><a href="#cb40-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Reshape to long and wide format</span></span>
<span id="cb40-34"><a href="#cb40-34" aria-hidden="true" tabindex="-1"></a>long_summary <span class="ot">&lt;-</span> summary_df <span class="sc">%&gt;%</span></span>
<span id="cb40-35"><a href="#cb40-35" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>samples, <span class="at">names_to =</span> <span class="st">"Sample"</span>, <span class="at">values_to =</span> <span class="st">"Expression"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb40-36"><a href="#cb40-36" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">Receptor =</span> samples)</span>
<span id="cb40-37"><a href="#cb40-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-38"><a href="#cb40-38" aria-hidden="true" tabindex="-1"></a>df_wide <span class="ot">&lt;-</span> long_summary <span class="sc">%&gt;%</span></span>
<span id="cb40-39"><a href="#cb40-39" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> Receptor, <span class="at">values_from =</span> Expression)</span>
<span id="cb40-40"><a href="#cb40-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-41"><a href="#cb40-41" aria-hidden="true" tabindex="-1"></a>df_wide <span class="ot">&lt;-</span> df_wide <span class="sc">%&gt;%</span></span>
<span id="cb40-42"><a href="#cb40-42" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Difference =</span> <span class="sc">!!</span><span class="fu">sym</span>(treatment_name) <span class="sc">-</span> <span class="sc">!!</span><span class="fu">sym</span>(control1_name) <span class="sc">-</span> <span class="sc">!!</span><span class="fu">sym</span>(control2_name))</span>
<span id="cb40-43"><a href="#cb40-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-44"><a href="#cb40-44" aria-hidden="true" tabindex="-1"></a>df_wide_pos <span class="ot">&lt;-</span> df_wide <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(Difference <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb40-45"><a href="#cb40-45" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(<span class="fu">desc</span>(Difference)) </span>
<span id="cb40-46"><a href="#cb40-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-47"><a href="#cb40-47" aria-hidden="true" tabindex="-1"></a>df_wide_pos <span class="ot">&lt;-</span> <span class="fu">head</span>(df_wide_pos, <span class="dv">50</span>)</span>
<span id="cb40-48"><a href="#cb40-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-49"><a href="#cb40-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a total expression column for sorting</span></span>
<span id="cb40-50"><a href="#cb40-50" aria-hidden="true" tabindex="-1"></a>df_wide_pos <span class="ot">&lt;-</span> df_wide_pos <span class="sc">%&gt;%</span></span>
<span id="cb40-51"><a href="#cb40-51" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Total_expression =</span> <span class="sc">!!</span><span class="fu">sym</span>(treatment_name) <span class="sc">+</span> <span class="sc">!!</span><span class="fu">sym</span>(control1_name) <span class="sc">+</span> <span class="sc">!!</span><span class="fu">sym</span>(control2_name)) <span class="sc">%&gt;%</span></span>
<span id="cb40-52"><a href="#cb40-52" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(Total_expression)</span>
<span id="cb40-53"><a href="#cb40-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-54"><a href="#cb40-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Preserve this order in the plot by setting Sample as a factor</span></span>
<span id="cb40-55"><a href="#cb40-55" aria-hidden="true" tabindex="-1"></a>df_wide_pos<span class="sc">$</span>Sample <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_wide_pos<span class="sc">$</span>Sample, <span class="at">levels =</span> df_wide_pos<span class="sc">$</span>Sample)</span>
<span id="cb40-56"><a href="#cb40-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-57"><a href="#cb40-57" aria-hidden="true" tabindex="-1"></a>df_long_pos <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb40-58"><a href="#cb40-58" aria-hidden="true" tabindex="-1"></a>  df_wide_pos,</span>
<span id="cb40-59"><a href="#cb40-59" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"Difference"</span>, control1_name, control2_name),</span>
<span id="cb40-60"><a href="#cb40-60" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to =</span> <span class="st">"Orthogonal Receptor"</span>,</span>
<span id="cb40-61"><a href="#cb40-61" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to =</span> <span class="st">"Expression"</span></span>
<span id="cb40-62"><a href="#cb40-62" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-63"><a href="#cb40-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-64"><a href="#cb40-64" aria-hidden="true" tabindex="-1"></a><span class="co"># Set plotting order</span></span>
<span id="cb40-65"><a href="#cb40-65" aria-hidden="true" tabindex="-1"></a>df_long_pos<span class="sc">$</span><span class="st">`</span><span class="at">Orthogonal Receptor</span><span class="st">`</span> <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_long_pos<span class="sc">$</span><span class="st">`</span><span class="at">Orthogonal Receptor</span><span class="st">`</span>, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Difference"</span>, control1_name, control2_name))</span>
<span id="cb40-66"><a href="#cb40-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-67"><a href="#cb40-67" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(demo_data, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">layer =</span> <span class="st">"counts"</span>)</span>
<span id="cb40-68"><a href="#cb40-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-69"><a href="#cb40-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming you already have counts</span></span>
<span id="cb40-70"><a href="#cb40-70" aria-hidden="true" tabindex="-1"></a>ensembl_ids <span class="ot">&lt;-</span> <span class="fu">rownames</span>(counts)</span>
<span id="cb40-71"><a href="#cb40-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-72"><a href="#cb40-72" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb40-73"><a href="#cb40-73" aria-hidden="true" tabindex="-1"></a>mart <span class="ot">&lt;-</span> <span class="fu">useMart</span>(<span class="st">"ensembl"</span>, <span class="at">dataset =</span> <span class="st">"mmusculus_gene_ensembl"</span>)</span>
<span id="cb40-74"><a href="#cb40-74" aria-hidden="true" tabindex="-1"></a>gene_map <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb40-75"><a href="#cb40-75" aria-hidden="true" tabindex="-1"></a>  <span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"mgi_symbol"</span>),</span>
<span id="cb40-76"><a href="#cb40-76" aria-hidden="true" tabindex="-1"></a>  <span class="at">filters =</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb40-77"><a href="#cb40-77" aria-hidden="true" tabindex="-1"></a>  <span class="at">values =</span> ensembl_ids,</span>
<span id="cb40-78"><a href="#cb40-78" aria-hidden="true" tabindex="-1"></a>  <span class="at">mart =</span> mart</span>
<span id="cb40-79"><a href="#cb40-79" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-80"><a href="#cb40-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-81"><a href="#cb40-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure gene_map has no duplicated Ensembl IDs</span></span>
<span id="cb40-82"><a href="#cb40-82" aria-hidden="true" tabindex="-1"></a>gene_map_unique <span class="ot">&lt;-</span> gene_map[<span class="sc">!</span><span class="fu">duplicated</span>(gene_map<span class="sc">$</span>ensembl_gene_id), ]</span>
<span id="cb40-83"><a href="#cb40-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-84"><a href="#cb40-84" aria-hidden="true" tabindex="-1"></a><span class="co"># Map Sample column using gene_map</span></span>
<span id="cb40-85"><a href="#cb40-85" aria-hidden="true" tabindex="-1"></a>df_long_pos<span class="sc">$</span>Sample <span class="ot">&lt;-</span> gene_map_unique<span class="sc">$</span>mgi_symbol[<span class="fu">match</span>(df_long_pos<span class="sc">$</span>Sample, gene_map_unique<span class="sc">$</span>ensembl_gene_id)]</span>
<span id="cb40-86"><a href="#cb40-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-87"><a href="#cb40-87" aria-hidden="true" tabindex="-1"></a><span class="co"># If some MGI symbols are NA, fall back to Ensembl ID</span></span>
<span id="cb40-88"><a href="#cb40-88" aria-hidden="true" tabindex="-1"></a>df_long_pos<span class="sc">$</span>Sample[<span class="fu">is.na</span>(df_long_pos<span class="sc">$</span>Sample)] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(df_long_pos<span class="sc">$</span>Sample[<span class="fu">is.na</span>(df_long_pos<span class="sc">$</span>Sample)])</span>
<span id="cb40-89"><a href="#cb40-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-90"><a href="#cb40-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Set color scheme with evaluated names</span></span>
<span id="cb40-91"><a href="#cb40-91" aria-hidden="true" tabindex="-1"></a>custom_colors <span class="ot">&lt;-</span> <span class="fu">setNames</span>(</span>
<span id="cb40-92"><a href="#cb40-92" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"#D8B4E2"</span>, <span class="st">"#AED9E0"</span>, <span class="st">"#F8C8DC"</span>),</span>
<span id="cb40-93"><a href="#cb40-93" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"Difference"</span>, control1_name, control2_name)</span>
<span id="cb40-94"><a href="#cb40-94" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-95"><a href="#cb40-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-96"><a href="#cb40-96" aria-hidden="true" tabindex="-1"></a>df_long_pos <span class="ot">&lt;-</span> df_long_pos <span class="sc">%&gt;%</span></span>
<span id="cb40-97"><a href="#cb40-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Sample) <span class="sc">&amp;</span> Sample <span class="sc">!=</span> <span class="st">""</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">Orthogonal Receptor</span><span class="st">`</span>) <span class="sc">&amp;</span> <span class="st">`</span><span class="at">Orthogonal Receptor</span><span class="st">`</span> <span class="sc">!=</span> <span class="st">""</span>)</span>
<span id="cb40-98"><a href="#cb40-98" aria-hidden="true" tabindex="-1"></a>df_long_pos <span class="ot">&lt;-</span> df_long_pos <span class="sc">%&gt;%</span></span>
<span id="cb40-99"><a href="#cb40-99" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"^Gm|^</span><span class="sc">\\</span><span class="st">d{4}"</span>, Sample))</span>
<span id="cb40-100"><a href="#cb40-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-101"><a href="#cb40-101" aria-hidden="true" tabindex="-1"></a>pos_synergy_plot <span class="ot">&lt;-</span> <span class="fu">plot_synergy_barplot</span>(df_long_pos, treatment_name, custom_colors, <span class="at">sample_col =</span> <span class="st">"Sample"</span>,</span>
<span id="cb40-102"><a href="#cb40-102" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression_col =</span> <span class="st">"Expression"</span>,</span>
<span id="cb40-103"><a href="#cb40-103" aria-hidden="true" tabindex="-1"></a>  <span class="at">cytokine_col =</span> <span class="st">"Orthogonal Receptor"</span>)</span>
<span id="cb40-104"><a href="#cb40-104" aria-hidden="true" tabindex="-1"></a>pos_synergy_plot</span>
<span id="cb40-105"><a href="#cb40-105" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(pos_synergy_plot)</span>
<span id="cb40-106"><a href="#cb40-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-107"><a href="#cb40-107" aria-hidden="true" tabindex="-1"></a><span class="fu">save_plot_as_png</span>(pos_synergy_plot, <span class="fu">paste0</span>(treatment, <span class="st">"_difference.png"</span>))</span>
<span id="cb40-108"><a href="#cb40-108" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using an external vector in selections was deprecated in tidyselect 1.1.0.
ℹ Please use `all_of()` or `any_of()` instead.
  # Was:
  data %&gt;% select(control1_name)

  # Now:
  data %&gt;% select(all_of(control1_name))

See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using an external vector in selections was deprecated in tidyselect 1.1.0.
ℹ Please use `all_of()` or `any_of()` instead.
  # Was:
  data %&gt;% select(control2_name)

  # Now:
  data %&gt;% select(all_of(control2_name))

See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="06_synergy_demo_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="06_synergy_demo_files/figure-html/unnamed-chunk-15-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="why-this-matters-for-o10r-biology" class="level3">
<h3 class="anchored" data-anchor-id="why-this-matters-for-o10r-biology">Why this matters for o10R biology</h3>
<ul>
<li><p><strong>o10R phenotype:</strong>&nbsp;IL-10 pathway rewiring can promote a&nbsp;<strong>fit, persistence-prone</strong>&nbsp;state while retaining effector potential. In our pipeline, this is reflected by&nbsp;<strong>exclusive/unique DE genes</strong>&nbsp;under o10R contexts and&nbsp;<strong>positive “Difference” bars</strong>&nbsp;in the combo plots.</p></li>
<li><p><strong>STAT programs:</strong>&nbsp;Although IL-10 canonically engages&nbsp;<strong>STAT3</strong>, crosstalk can broaden to&nbsp;<strong>STAT1/5</strong>&nbsp;in engineered contexts. The DE and synergy lists provide concrete gene-level evidence to probe these axes (e.g., STAT target signatures, survival/exhaustion modules).</p></li>
<li><p><strong>Translational angle:</strong>&nbsp;Efficacy&nbsp;<strong>without preconditioning</strong>&nbsp;is consistent with durable transcriptional programs rather than transient boosts; synergy genes are natural candidates for&nbsp;<strong>pathway enrichment</strong>&nbsp;and&nbsp;<strong>mechanistic validation</strong>&nbsp;(e.g., knockdown/overexpression, phospho-flow).</p></li>
</ul>
</section>
<section id="how-to-read-the-outputs" class="level3">
<h3 class="anchored" data-anchor-id="how-to-read-the-outputs">How to read the outputs</h3>
<ul>
<li><p><code>padj_005/*_upsig.csv</code>&nbsp;—&nbsp;<strong>Condition-specific upregulated</strong>&nbsp;genes vs 10R (use these for pathway enrichment per condition).</p></li>
<li><p><code>synergy/&lt;pair&gt;/analysis_results/common_genes.csv</code>&nbsp;— shared signature across files for that pair.</p></li>
<li><p><code>.../exclusive_genes_*.csv</code>&nbsp;—&nbsp;<strong>most characteristic</strong>&nbsp;genes for a file/condition.</p></li>
<li><p><code>&lt;pair&gt;_difference.png</code>&nbsp;—&nbsp;<strong>Supra-additive</strong>&nbsp;genes rank-ordered; taller&nbsp;<strong>Difference</strong>&nbsp;bars = stronger synergy beyond the sum of singles.</p></li>
</ul>
</section>
<section id="next-suggested-checks-optional-to-add-later" class="level3">
<h3 class="anchored" data-anchor-id="next-suggested-checks-optional-to-add-later">Next suggested checks (optional to add later)</h3>
<ul>
<li><p><strong>Cell-type stratification / pseudobulk</strong>&nbsp;to reduce cell-mixing effects.</p></li>
<li><p><strong>Robust summaries</strong>&nbsp;(median or trimmed mean) alongside mean-based synergy.</p></li>
<li><p><strong>Pathway GSEA</strong>&nbsp;on synergy genes (STAT modules, survival/metabolism, exhaustion).</p></li>
<li><p><strong>Model-based interaction</strong>&nbsp;(GLM with interaction term) to corroborate the rule-based Z &gt; X + Y calls.</p></li>
</ul>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>