<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>07_boxplot_demo</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="07_boxplots_demo_files/libs/clipboard/clipboard.min.js"></script>
<script src="07_boxplots_demo_files/libs/quarto-html/quarto.js"></script>
<script src="07_boxplots_demo_files/libs/quarto-html/popper.min.js"></script>
<script src="07_boxplots_demo_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="07_boxplots_demo_files/libs/quarto-html/anchor.min.js"></script>
<link href="07_boxplots_demo_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="07_boxplots_demo_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="07_boxplots_demo_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="07_boxplots_demo_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="07_boxplots_demo_files/libs/bootstrap/bootstrap-c0367b04c37547644fece4185067e4a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">07_boxplot_demo</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(dplyr))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(purrr))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tibble))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(SeuratObject))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(Seurat))</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(patchwork))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(ggplot2))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(readxl))</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(plyr))</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to save plot as PNG</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>save_plot_as_png <span class="ot">&lt;-</span> <span class="cf">function</span>(plot, filename) {</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">png</span>(filename, <span class="at">res =</span> <span class="dv">250</span>, <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">2000</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="why-was-it-done" class="level1">
<h1>Why was it done?</h1>
<p>Solid-tumor T-cell engineering often yields&nbsp;<strong>condition-specific</strong>&nbsp;transcriptional effects. You wanted a fast, reproducible screen to (1)&nbsp;<strong>flag genes that are selectively elevated</strong>&nbsp;in one sample/condition relative to all others, and (2)&nbsp;<strong>visualize</strong>those differences at whole-dataset and per-cell-type levels with simple stats. This helps prioritize candidate markers/pathways for follow-up (e.g., enrichment, synergy checks, validation).</p>
</section>
<section id="objective" class="level1">
<h1>Objective</h1>
<ul>
<li><p><strong>Primary:</strong>&nbsp;Identify genes whose mean expression is&nbsp;<strong>substantially higher in one sample</strong>&nbsp;than the average of the remaining samples, and&nbsp;<strong>summarize/visualize</strong>&nbsp;those genes overall and by cell type.</p></li>
<li><p><strong>Operational thresholds:</strong>&nbsp;keep genes expressed in ≥5% of cells; call a gene “selectively high” if its fold change (FC) in any sample vs.&nbsp;the mean of others is&nbsp;<strong>&gt; 2.5</strong>.</p></li>
</ul>
</section>
<section id="summary-what-the-pipeline-does" class="level1">
<h1>Summary (what the pipeline does)</h1>
<ol type="1">
<li><p>Compute&nbsp;<strong>per-sample average expression</strong>&nbsp;for all genes (RNA assay).</p></li>
<li><p>Filter to genes expressed in&nbsp;<strong>≥5%</strong>&nbsp;of cells.</p></li>
<li><p>For each gene, compute a&nbsp;<strong>sample-wise FC</strong>&nbsp;= (sample mean) / (mean of other samples).</p></li>
<li><p>Keep genes with&nbsp;<strong>FC &gt; 2.5</strong>&nbsp;in&nbsp;<em>any</em>&nbsp;sample ⇒ write to&nbsp;<code>preliminary/significant_foldchange_genes.csv</code>.</p></li>
<li><p>For each retained gene, generate:</p>
<ul>
<li><p><strong>All-cells boxplots</strong>&nbsp;across samples with pairwise&nbsp;<strong>Wilcoxon</strong>&nbsp;p-value annotations.</p></li>
<li><p><strong>Per-cell-type boxplots</strong>&nbsp;(same tests within each cell type).</p></li>
<li><p><strong>Combined panels</strong>&nbsp;(all-cells + each cell type) via cowplot.</p></li>
</ul></li>
<li><p>Save plots under&nbsp;<code>plots/all_cells/</code>,&nbsp;<code>plots/per_celltype/&lt;celltype&gt;/</code>, and&nbsp;<code>plots/cowplot_combined/</code>.</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load("../data/demo_data_annotated_nobiomart.RData") </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#loading this later</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/demo_data_annotated_biomart.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="synergy-genes-table" class="level1">
<h1>Synergy Genes Table</h1>
<p>from 06_synergy_demo.qmd I saved the following</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv2</span>(exclusive_genes_list_synergy<span class="sc">$</span><span class="st">`</span><span class="at">4R_12R</span><span class="st">`</span>,<span class="st">"4R_12R_exclusive_genes_list_synergy.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="and-now-i-will-read" class="level1">
<h1>and Now i will read</h1>
<p>synergy_genes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>synergy_genes_table <span class="ot">&lt;-</span> <span class="fu">read.csv2</span>(<span class="st">"4R_12R_exclusive_genes_list_synergy.csv"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="st">"gene_name"</span> <span class="sc">%in%</span> <span class="fu">names</span>(synergy_genes_table)) {</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(synergy_genes_table)[<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">"gene_name"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>synergy_genes_table<span class="sc">$</span>gene_name <span class="ot">&lt;-</span> <span class="fu">trimws</span>(synergy_genes_table<span class="sc">$</span>gene_name)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(synergy_genes_table,<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    X          gene_name
1   1 ENSMUSG00000130707
2   2 ENSMUSG00000138022
3   3 ENSMUSG00000087641
4   4 ENSMUSG00000137281
5   5 ENSMUSG00000086686
6   6 ENSMUSG00000132261
7   7 ENSMUSG00000021804
8   8 ENSMUSG00000038859
9   9 ENSMUSG00000003436
10 10 ENSMUSG00000038599
11 11 ENSMUSG00000143050
12 12 ENSMUSG00000103692
13 13 ENSMUSG00000127260
14 14 ENSMUSG00000039714
15 15 ENSMUSG00000118454
16 16 ENSMUSG00000031766
17 17 ENSMUSG00000037196
18 18 ENSMUSG00000138369
19 19 ENSMUSG00000117356
20 20 ENSMUSG00000019817</code></pre>
</div>
</div>
<p>converting using biomart</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#counts &lt;- GetAssayData(demo_data, assay = "RNA", layer = "counts")</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>ensembl_ids <span class="ot">&lt;-</span> synergy_genes_table<span class="sc">$</span>gene_name</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>mart <span class="ot">&lt;-</span> <span class="fu">useMart</span>(<span class="st">"ensembl"</span>, <span class="at">dataset =</span> <span class="st">"mmusculus_gene_ensembl"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>gene_map <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"mgi_symbol"</span>),</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">filters =</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">values =</span> ensembl_ids,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">mart =</span> mart</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>gene_map_unique <span class="ot">&lt;-</span> gene_map[<span class="sc">!</span><span class="fu">duplicated</span>(gene_map<span class="sc">$</span>ensembl_gene_id), ]</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gene_map_unique,<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      ensembl_gene_id mgi_symbol
1  ENSMUSG00000003436       Dll3
2  ENSMUSG00000005493       Msh4
3  ENSMUSG00000018543     Spmap1
4  ENSMUSG00000019817     Plagl1
5  ENSMUSG00000020020      Usp44
6  ENSMUSG00000020785     Camkk1
7  ENSMUSG00000021804        Rgr
8  ENSMUSG00000022540      Rogdi
9  ENSMUSG00000024593     Megf10
10 ENSMUSG00000024664      Fads3
11 ENSMUSG00000024912      Fosl1
12 ENSMUSG00000026271      Gpr35
13 ENSMUSG00000026858      Miga2
14 ENSMUSG00000027068      Dhrs9
15 ENSMUSG00000027690     Slc2a2
16 ENSMUSG00000027932    Slc27a3
17 ENSMUSG00000030499     Kctd15
18 ENSMUSG00000030717      Nupr1
19 ENSMUSG00000031766    Slc12a3
20 ENSMUSG00000032372     Plscr2</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>synergy_genes_table<span class="sc">$</span>mapped_gene_name <span class="ot">&lt;-</span> gene_map_unique<span class="sc">$</span>mgi_symbol[<span class="fu">match</span>(synergy_genes_table<span class="sc">$</span>gene_name, gene_map_unique<span class="sc">$</span>ensembl_gene_id)]</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>synergy_genes_table<span class="sc">$</span>mapped_gene_name[<span class="fu">is.na</span>(synergy_genes_table<span class="sc">$</span>mapped_gene_name)] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(synergy_genes_table<span class="sc">$</span>mapped_gene_name[<span class="fu">is.na</span>(synergy_genes_table<span class="sc">$</span>mapped_gene_name)])</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(synergy_genes_table,<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    X          gene_name mapped_gene_name
1   1 ENSMUSG00000130707          Gm62520
2   2 ENSMUSG00000138022          Gm67435
3   3 ENSMUSG00000087641          Gm12811
4   4 ENSMUSG00000137281          Gm59254
5   5 ENSMUSG00000086686    F630206G17Rik
6   6 ENSMUSG00000132261          Gm65255
7   7 ENSMUSG00000021804              Rgr
8   8 ENSMUSG00000038859         Baiap2l1
9   9 ENSMUSG00000003436             Dll3
10 10 ENSMUSG00000038599            Capn8
11 11 ENSMUSG00000143050          Gm61331
12 12 ENSMUSG00000103692    4930503O07Rik
13 13 ENSMUSG00000127260          Gm72467
14 14 ENSMUSG00000039714            Cplx3
15 15 ENSMUSG00000118454            Wdr88
16 16 ENSMUSG00000031766          Slc12a3
17 17 ENSMUSG00000037196            Pacrg
18 18 ENSMUSG00000138369          Gm32881
19 19 ENSMUSG00000117356          Gm46633
20 20 ENSMUSG00000019817           Plagl1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>synergy_genes <span class="ot">&lt;-</span> synergy_genes_table<span class="sc">$</span>mapped_gene_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="methods-code-analysis-steps" class="level1">
<h1>Methods (code → analysis steps)</h1>
<ul>
<li><p><strong>Data setup:</strong>&nbsp;<code>DefaultAssay(demo_data) &lt;- "RNA"</code>. Use&nbsp;<code>AverageExpression(..., group.by = "samples")</code>to get a gene × sample matrix of means.</p></li>
<li><p><strong>Expression filter:</strong>&nbsp;<code>min_cells = 0.05 * ncol(demo_data)</code>; keep genes with counts&gt;0 in ≥ min_cells (<code>Matrix::rowSums(...)</code>).</p></li>
<li><p><strong>Per-sample fold change:</strong>&nbsp;For each gene row, loop over samples&nbsp;<em>i</em>&nbsp;and compute&nbsp;<code>this_val / mean(other_vals)</code>, with&nbsp;<code>1e-6</code>&nbsp;for numerical stability.</p></li>
<li><p><strong>Hit list:</strong>&nbsp;select genes with FC&gt;2.5 in any column.</p></li>
<li><p><strong>Statistics &amp; plotting:</strong></p>
<ul>
<li><p>Build long data frames via&nbsp;<code>FetchData(..., vars = c(gene, "samples"))</code>.</p></li>
<li><p><strong>Boxplots</strong>&nbsp;(no outliers drawn) +&nbsp;<strong>pairwise Wilcoxon</strong>&nbsp;via&nbsp;<code>ggpubr::stat_compare_means</code>&nbsp;across all sample pairs.</p></li>
<li><p>Ensure sample ordering (places&nbsp;<code>"box"</code>&nbsp;first in the legend/axis if present).</p></li>
<li><p><strong>Per-cell-type</strong>&nbsp;subsets using&nbsp;<code>metadata$cell_annotation</code>&nbsp;to stratify.</p></li>
<li><p>Combine panels with&nbsp;<code>cowplot::plot_grid</code>, save PNGs (300 dpi).</p></li>
</ul></li>
</ul>
</section>
<section id="tasks-automated" class="level1">
<h1>Tasks automated</h1>
<ul>
<li><p>Gene-level&nbsp;<strong>screening</strong>&nbsp;for selective up-regulation by sample.</p></li>
<li><p>Batch&nbsp;<strong>plot generation</strong>&nbsp;(overall and per cell type) with inline&nbsp;<strong>p-value labels</strong>.</p></li>
<li><p>Structured&nbsp;<strong>file outputs</strong>&nbsp;for easy review and downstream use.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># genes, samples, conditions.</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Set default assay</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(demo_data) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Compute average gene expression per sample</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>avg_exp <span class="ot">&lt;-</span> <span class="fu">AverageExpression</span>(demo_data, <span class="at">group.by =</span> <span class="st">"samples"</span>, <span class="at">return.seurat =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>RNA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.
First group.by variable `samples` starts with a number, appending `g` to ensure valid variable names
This message is displayed once per session.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>min_cells <span class="ot">&lt;-</span> <span class="fl">0.02</span> <span class="sc">*</span> <span class="fu">ncol</span>(demo_data)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>keep_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(demo_data)[Matrix<span class="sc">::</span><span class="fu">rowSums</span>(<span class="fu">GetAssayData</span>(demo_data, <span class="at">slot =</span> <span class="st">"counts"</span>) <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> min_cells]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
ℹ Please use the `layer` argument instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset average expression</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>avg_exp_filtered <span class="ot">&lt;-</span> avg_exp[<span class="fu">rownames</span>(avg_exp) <span class="sc">%in%</span> keep_genes, ]</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Compute fold change of each gene in each sample compared to mean of other samples</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>fc_matrix <span class="ot">&lt;-</span> <span class="fu">apply</span>(avg_exp_filtered , <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="fu">seq_along</span>(x), <span class="cf">function</span>(i) {</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    this_val <span class="ot">&lt;-</span> x[i]</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    other_avg <span class="ot">&lt;-</span> <span class="fu">mean</span>(x[<span class="sc">-</span>i])</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(this_val <span class="sc">/</span> (other_avg <span class="sc">+</span> <span class="fl">1e-6</span>))  <span class="co"># avoid division by zero</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 3: Transpose and convert to dataframe</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>fc_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(fc_matrix))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(fc_df) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(avg_exp_filtered )</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>fc_df<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(avg_exp_filtered)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 4: Filter genes with fold change &gt; 1.5 in any sample</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>sig_genes <span class="ot">&lt;-</span> fc_df[<span class="fu">rowSums</span>(fc_df[, <span class="sc">-</span><span class="fu">ncol</span>(fc_df)] <span class="sc">&gt;</span> <span class="fl">2.5</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="co"># these exclusive_genes_list_synergy is from the 06_synergy_demo.qmd</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>genes_in_common <span class="ot">&lt;-</span> sig_genes <span class="sc">%&gt;%</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a> dplyr<span class="sc">::</span><span class="fu">filter</span>(gene <span class="sc">%in%</span> synergy_genes)</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>genes_in_common</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              g12R      g10R   g12R+4R       g4R    g4R+12R        box     gene
Lif      0.7656546 0.8985252 1.2672007 0.6060122  3.3968289 0.00000000      Lif
Baiap2l1 0.2316152 0.2855069 0.1583805 0.4254057 10.6891688 0.62586535 Baiap2l1
Plscr1l1 0.6399589 1.0716134 0.8269767 0.9440208  3.2128970 0.09178715 Plscr1l1
Rpph1    0.4230989 0.7828143 0.6821072 0.7624555  2.5190741 1.24397212    Rpph1
Gm20619  0.2788455 0.4394929 0.9492093 0.2478243  0.9552331 4.98367770  Gm20619
Gm60415  0.7964241 0.6609338 0.9075342 1.2369499  2.6014119 0.27211653  Gm60415
Gm75556  0.3984481 1.0137197 1.6193980 0.1583538  3.8049836 0.26394266  Gm75556
Gm61125  0.5104201 0.6566904 1.0921187 0.6031904  2.7020252 0.90693215  Gm61125</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># write.csv(sig_genes, "significant_foldchange_genes.csv", row.names = FALSE)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ggpubr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:plyr':

    mutate</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ---------------- Setup ----------------</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(demo_data) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>expr_matrix <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(demo_data, <span class="at">slot =</span> <span class="st">"data"</span>)  <span class="co"># assumes NormalizeData done</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> demo_data<span class="sc">@</span>meta.data</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Get a CHARACTER VECTOR of gene names (not rows of df)</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> genes_in_common<span class="sc">$</span>gene</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">unique</span>(genes_in_common<span class="sc">$</span>gene), <span class="dv">4</span>)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>celltypes <span class="ot">&lt;-</span> <span class="fu">unique</span>(metadata<span class="sc">$</span>cell_annotation)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Helper: boxplot</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>plot_box <span class="ot">&lt;-</span> <span class="cf">function</span>(df, gene, title_prefix) {</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>  boxplot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">y =</span> expression, <span class="at">fill =</span> samples)) <span class="sc">+</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>, <span class="at">width =</span> <span class="fl">0.6</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_compare_means</span>(</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">method =</span> <span class="st">"wilcox.test"</span>,</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">comparisons =</span> <span class="fu">combn</span>(<span class="fu">levels</span>(<span class="fu">factor</span>(df<span class="sc">$</span>samples)), <span class="dv">2</span>, <span class="at">simplify =</span> <span class="cn">FALSE</span>),</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">label =</span> <span class="st">"p.signif"</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">13</span>) <span class="sc">+</span></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="fu">paste</span>(title_prefix, gene)) <span class="sc">+</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Expression Level"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Sample"</span>)</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(boxplot1)</span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create base directories</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"plots/all_cells"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"plots/per_celltype"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ---------------- General Boxplots (All Cells) ----------------</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"📤 Saving all-cells boxplots...</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>📤 Saving all-cells boxplots...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>n_all <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (g <span class="cf">in</span> top_genes) {</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># scalar checks only</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>(g <span class="sc">%in%</span> <span class="fu">rownames</span>(expr_matrix))) <span class="cf">next</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  file_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"plots/all_cells"</span>, <span class="fu">paste0</span>(g, <span class="st">"_boxplot.png"</span>))</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">file.exists</span>(file_path)) <span class="cf">next</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># use the SAME object you prepared (demo_data)</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(demo_data, <span class="at">vars =</span> <span class="fu">c</span>(g, <span class="st">"samples"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"expression"</span>, <span class="st">"samples"</span>))</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">plot_box</span>(df, g, <span class="at">title_prefix =</span> <span class="st">"All Cells — "</span>)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">filename =</span> file_path, <span class="at">plot =</span> p, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">5</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  n_all <span class="ot">&lt;-</span> n_all <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="07_boxplots_demo_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"✅ New all-cell plots saved:"</span>, n_all, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>✅ New all-cell plots saved: 4 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ---------------- Per-Celltype Boxplots ----------------</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"📤 Saving per-celltype boxplots...</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>📤 Saving per-celltype boxplots...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>n_ct <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (g <span class="cf">in</span> top_genes) {</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>(g <span class="sc">%in%</span> <span class="fu">rownames</span>(expr_matrix))) <span class="cf">next</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (ct <span class="cf">in</span> celltypes) {</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>    out_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"plots/per_celltype"</span>, <span class="fu">gsub</span>(<span class="st">"[/ ]"</span>, <span class="st">"_"</span>, ct))</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(out_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>    file_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(out_dir, <span class="fu">paste0</span>(g, <span class="st">"_"</span>, <span class="fu">gsub</span>(<span class="st">"[/ ]"</span>, <span class="st">"_"</span>, ct), <span class="st">".png"</span>))</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">file.exists</span>(file_path)) <span class="cf">next</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>    cell_ids <span class="ot">&lt;-</span> <span class="fu">rownames</span>(metadata)[metadata<span class="sc">$</span>cell_annotation <span class="sc">==</span> ct]</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>    cell_ids <span class="ot">&lt;-</span> <span class="fu">intersect</span>(cell_ids, <span class="fu">colnames</span>(expr_matrix))</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(cell_ids) <span class="sc">&lt;</span> <span class="dv">2</span>) <span class="cf">next</span>  <span class="co"># need ≥2 for boxplot + stats</span></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">expression =</span> <span class="fu">as.numeric</span>(expr_matrix[g, cell_ids]),</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">samples    =</span> metadata[cell_ids, <span class="st">"samples"</span>, <span class="at">drop =</span> <span class="cn">TRUE</span>],</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="fu">plot_box</span>(df, g, <span class="at">title_prefix =</span> <span class="fu">paste</span>(<span class="st">"Cell Type:"</span>, ct, <span class="st">"—"</span>))</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggsave</span>(<span class="at">filename =</span> file_path, <span class="at">plot =</span> p, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">5</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>    n_ct <span class="ot">&lt;-</span> n_ct <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0.707499905428215, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0,
0.430838115581402, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.707499905428215, 0, 0), c(0, 0, 0,
0.430838115581402, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.707499905428215, 0, 0), c(0, 0, 0, 0), :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.430838115581402, 1.40310190551928,
: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.164769744086739, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0), paired = FALSE):
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0), c(0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0,
0.238444478044458, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0.238444478044458, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0), paired = FALSE):
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.238444478044458, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.692298262878942, 0, 0, 0, 0,
0.582851864616271: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.692298262878942, 0, 0, 0, 0,
0.582851864616271: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.692298262878942, 0, 0, 0, 0,
0.582851864616271: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(1.40774693010508, 0, 0), c(0, 0,
0.441912131848367, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(1.40774693010508, 0, 0), c(0, 0, 0, 0), :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0.441912131848367, 0.592121530583344, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.164769744086739, 0.429550735904047,
: cannot compute exact p-value with ties</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.164769744086739, 0.429550735904047,
: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="07_boxplots_demo_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"✅ New per-celltype plots saved:"</span>, n_ct, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>✅ New per-celltype plots saved: 28 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'cowplot'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ggpubr':

    get_legend</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:patchwork':

    align_plots</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"plots/cowplot_combined"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over genes</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gene <span class="cf">in</span> top_genes) {</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>(gene <span class="sc">%in%</span> <span class="fu">rownames</span>(expr_matrix))) <span class="cf">next</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  plots <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># All cells plot</span></span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>  df_all <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(demo_data, <span class="at">vars =</span> <span class="fu">c</span>(gene, <span class="st">"samples"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"expression"</span>, <span class="st">"samples"</span>))</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>  plots[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">plot_box</span>(df_all, gene, <span class="at">title_prefix =</span> <span class="st">"All Cells — "</span>)</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Per-celltype plots</span></span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (ct <span class="cf">in</span> celltypes) {</span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>    cell_ids <span class="ot">&lt;-</span> <span class="fu">rownames</span>(metadata <span class="sc">%&gt;%</span> <span class="fu">filter</span>(cell_annotation <span class="sc">==</span> ct))</span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>    expr_values <span class="ot">&lt;-</span> expr_matrix[gene, cell_ids]</span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">expression =</span> <span class="fu">as.numeric</span>(expr_values),</span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">samples =</span> metadata[cell_ids, <span class="st">"samples"</span>]</span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>    plots[[<span class="fu">length</span>(plots) <span class="sc">+</span> <span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">plot_box</span>(df, gene, <span class="at">title_prefix =</span> <span class="fu">paste</span>(<span class="st">"Cell Type:"</span>, ct, <span class="st">"—"</span>))</span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine and save</span></span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a>  combined <span class="ot">&lt;-</span> cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plots, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">paste0</span>(<span class="st">"plots/cowplot_combined/"</span>, gene, <span class="st">"_combined.png"</span>),</span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot =</span> combined, <span class="at">width =</span> <span class="dv">14</span>, <span class="at">height =</span> <span class="dv">8</span>, <span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0.707499905428215, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0,
0.430838115581402, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.707499905428215, 0, 0), c(0, 0, 0,
0.430838115581402, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.707499905428215, 0, 0), c(0, 0, 0, 0), :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.430838115581402, 1.40310190551928,
: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.164769744086739, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0), paired = FALSE):
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0), c(0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0,
0.238444478044458, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0, 0, 0, 0), paired =
FALSE): cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0.238444478044458, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0), c(0, 0, 0, 0), paired = FALSE):
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.238444478044458, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.692298262878942, 0, 0, 0, 0,
0.582851864616271: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.692298262878942, 0, 0, 0, 0,
0.582851864616271: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0.692298262878942, 0, 0, 0, 0,
0.582851864616271: cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(1.40774693010508, 0, 0), c(0, 0,
0.441912131848367, : cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(1.40774693010508, 0, 0), c(0, 0, 0, 0), :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0.441912131848367, 0.592121530583344, :
cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0.164769744086739, 0.429550735904047,
: cannot compute exact p-value with ties</code></pre>
</div>
</div>
<pre><code>Warning in wilcox.test.default(c(0, 0, 0, 0, 0, 0), c(0.707499905428215,  :
  cannot compute exact p-value with ties

HAPPENS BECAUSE

Those warnings are from the Wilcoxon rank-sum test inside stat_compare_means().
They happen when your two groups have many tied values (e.g., lots of 0’s in scRNA-seq), so the test can’t compute the exact p-value and falls back to an asymptotic one.
This is normal for zero-inflated single-cell data and doesn’t invalidate the plot; it’s just telling you why it switched methods.</code></pre>
</section>
<section id="results" class="level1">
<h1>Results</h1>
<ul>
<li><p><strong>Hit count:</strong>&nbsp;<code>N</code>&nbsp;genes passed FC&gt;2.5 in ≥1 sample (see&nbsp;<code>significant_foldchange_genes.csv</code>).</p></li>
<li><p><strong>Top examples:</strong>&nbsp;e.g.,&nbsp;<code>GeneA</code>,&nbsp;<code>GeneB</code>,&nbsp;<code>GeneC</code>&nbsp;peak in&nbsp;<strong>Sample X</strong>;&nbsp;<code>GeneD</code>&nbsp;peaks in&nbsp;<strong>Sample Y</strong>.</p></li>
<li><p><strong>Per-cell-type specificity:</strong>&nbsp;In cell type&nbsp;<strong>T</strong>,&nbsp;<code>GeneA</code>&nbsp;remains highest in&nbsp;<strong>Sample X</strong>&nbsp;(Wilcoxon p &lt; 0.01 across comparisons), suggesting true condition-specific regulation rather than compositional effects.</p></li>
<li><p><strong>Figure references:</strong></p>
<ul>
<li><p>All-cells boxplots:&nbsp;<code>plots/all_cells/&lt;gene&gt;_boxplot.png</code></p></li>
<li><p>Per-cell-type boxplots:&nbsp;<code>plots/per_celltype/&lt;celltype&gt;/&lt;gene&gt;_&lt;celltype&gt;.png</code></p></li>
<li><p>Combined panels:&nbsp;<code>plots/cowplot_combined/&lt;gene&gt;_combined.png</code></p></li>
</ul></li>
</ul>
</section>
<section id="interpretation-caveats" class="level1">
<h1>Interpretation &amp; caveats</h1>
<ul>
<li><p><strong>What it tells you:</strong>&nbsp;a quick, interpretable list of&nbsp;<strong>condition-selective genes</strong>, with visual confirmation globally and within cell types.</p></li>
<li><p><strong>Caveats:</strong></p>
<ul>
<li><p>Using&nbsp;<strong>means</strong>&nbsp;on scRNA-seq can be influenced by zero inflation and rare high expressers; consider also&nbsp;<strong>median</strong>&nbsp;or&nbsp;<strong>trimmed mean</strong>&nbsp;and/or&nbsp;<strong>pseudobulk</strong>&nbsp;per sample.</p></li>
<li><p>Pairwise Wilcoxon across many sample pairs per gene introduces&nbsp;<strong>multiple-testing</strong>&nbsp;concerns—treat p-labels as screening hints, and confirm in a DE framework if you plan claims.</p></li>
<li><p>If “box” is a non-biological control/batch, keep it for QC but avoid over-interpreting biological contrasts against it.</p></li>
</ul></li>
</ul>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>